---
title: "R code for comparing different genome catalogs"
author: "Mantas Sereika"
date: "2025"
output:
  html_document: default
  pdf_document: default
---


### Load dependencies
```{r, include=TRUE, message=FALSE}
library(stringr)
library(stringi)



library(ggsankey)



library(lemon)
library(pals)

library(RColorBrewer)
library(tools)




library(dplyr)
library(tidyverse)
library(ggplot2)
library(ggbeeswarm)
library(ggpubr)
library(ggplotify)
library(grid)
library(gridExtra)
library(ComplexUpset)
```


### Load MFD catalog data
```{r, include=TRUE}
mfd_lr <- read.delim("../analysis/datasets/dataset_S3.tsv", sep="\t", header=T)
mfd_sr_drep <- read.delim("../data/catalogs/mfd-sr_drep.tsv", sep="\t", header=T)

mags_drep <- read.delim("../data/catalogs/catalogs_drep.tsv", sep="\t", header=T)
mags_stat <- read.delim("../data/catalogs/catalogs_mags.tsv", sep="\t", header=T)

df_cat <- read.delim("../data/catalogs/catalogs.tsv", sep="\t", header=T, check.names = FALSE)
```


### Plot presets
```{r, include=TRUE}
preset_gen <- list(theme_bw(),
                   theme(legend.position = "bottom",
                         legend.text=element_text(size=7),
                         legend.title = element_text(size=7),
                         axis.title.y = element_text(size = 7), 
                         axis.title.x = element_text(size = 7),
                         axis.text.y = element_text(size = 6),
                         axis.text.x = element_text(size = 6),
                         panel.grid.minor = element_line(size = 0.15),
                         panel.grid.major = element_line(size = 0.3)))

preset_gen2 <- list(preset_gen, scale_y_discrete(expand = c(0, 0)),
                    scale_x_discrete(guide = guide_axis(angle = 30), expand = c(0, 0)),
                    guides(fill = guide_colorbar(title.position = "left", title.vjust = 0.75)),
                    scale_fill_gradient2(midpoint = 40, limits=c(0,80), breaks=c(0,20,40,60,80), labels=c("0","20","40","60","80"),
                                         low = "#b2182b", mid= "#fee08b", high = "#3288bd", na.value = "#3288bd"))
```


### Plot table with genome catalog stats
```{r, include=TRUE}
colnames(df_cat) <- gsub(" ", "\n", colnames(df_cat))
colnames(df_cat) <- gsub("_", " ", colnames(df_cat))

tt <- ttheme_default(
      core = list(fg_params = list(hjust = ifelse(col(df_cat) == 1, 0.5, 0.5),
                                   fontface = ifelse(col(df_cat) == 1, "bold", "plain"),
                                   fontsize = 6.25),
                  bg_params = list(fill = NA, col = NA),
                  padding = unit(c(1.25, 4.25), "mm")),
      colhead = list(fg_params = list(hjust = 0.5, fontface = "bold", fontsize = 6),
                     bg_params = list(fill = NA, col = NA),
                     padding = unit(c(1.25, 4), "mm")))

table_grob <- tableGrob(df_cat, theme=tt, rows = NULL) %>%
              tab_add_hline(at.row = 1:2, row.side = "top", linewidth = 1.5) %>%
              tab_add_hline(at.row = nrow(df_cat)+1, row.side = "bottom", linewidth = 1.5) 

table_grob$widths[1] <- unit(10, "mm")

plot_table <- ggplot() + annotation_custom(table_grob) + theme_void()
plot_table
```


### Upset plot for shared species between catalogs
```{r, include=TRUE}
mags_drep_ <- merge(mags_stat,mags_drep,by="bin")

mags_drep_[,"MFD-LR"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="MFD-LR",]$cluster,1,0)

mags_drep_[,"MFD-SR"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="MFD-SR",]$cluster,1,0)

mags_drep_[,"SPIRE"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="SPIRE",]$cluster,1,0)

mags_drep_[,"RBG"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="RBG",]$cluster,1,0)

mags_drep_[,"GEM"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="GEM",]$cluster,1,0)

mags_drep_[,"SMAG"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="SMAG",]$cluster,1,0)

mags_drep_[,"TPMC"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="TPMC",]$cluster,1,0)

mags_drep_[,"OWC"] <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$group=="OWC",]$cluster,1,0)

mags_drep_$cluster_qual <- ifelse(mags_drep_$cluster %in% mags_drep_[mags_drep_$MAG_status=="HQ",]$cluster,"HQ","MQ")
mags_drep_ <- mags_drep_[!duplicated(mags_drep_$cluster),]

names <- c("OWC","MFD-SR","TPMC","MFD-LR","GEM","SMAG","RBG","SPIRE")
mags_drep_[names] = mags_drep_[names] == 1
mags_drep_ <- mags_drep_[, c("cluster","cluster_qual",names)]
mags_drep_$cluster_qual <- factor(mags_drep_$cluster_qual,levels = c("MQ","HQ"))

plot_upset <- ComplexUpset::upset(mags_drep_,names,sort_sets=FALSE, set_sizes=FALSE, height_ratio=0.33, min_size=200, intersections='all',
              matrix=(intersection_matrix(geom=geom_point(size=1.7))),
              themes=upset_modify_themes(list('intersections_matrix'=theme(text=element_text(size=8),plot.margin = unit(c(-3,0,-2,0), 'cm')))),
              base_annotations=list('Intersection size'=intersection_size(bar_number_threshold=1, counts=T, text=list(size=1.7, angle=0), mapping=aes(fill=cluster_qual)) +
              stat_summary(fun = function(x) {sum_x <- sum(x); if (sum_x >= 1000) {return(sum_x)} else {return(NA)}},
                           fun.min = function(x) {sum_x <- sum(x); if (sum_x >= 1000) {return(sum_x)} else {return(NA)}},
                           geom='text', position=position_stack(vjust = 0.7), size=1.5, color="white", aes(label=..y..)) +
              scale_fill_manual(values = c("MQ" = "#d95f02","HQ"="#1b9e77"), name="") + ylab("Species-level clusters") + 
              theme(axis.text.y = element_text(size = 6), axis.title.y = element_text(size = 7),  plot.margin = unit(c(0,0,-2,0), 'cm'),legend.position = "none"))) +
              theme(legend.position = "none",panel.grid.minor = element_line(size = 0.15), panel.grid.major = element_line(size = 0.3), panel.spacing = unit(0, "cm")) + 
              labs(x="")
plot_upset
```


### Plot MAG counts for catalogs 
```{r, include=TRUE}
mags_n_drep <- aggregate(mags_stat$bin,by=list(mags_stat$MAG_status,mags_stat$group), FUN=length)
mags_n_drep <- mags_n_drep[order(mags_n_drep$x,decreasing=TRUE),]
mags_n_drep$Group.1 <- factor(mags_n_drep$Group.1,levels = c("MQ","HQ"))
mags_n_drep$Group.2 <- factor(mags_n_drep$Group.2,levels = c("OWC","MFD-SR","TPMC","MFD-LR","GEM","SMAG","RBG","SPIRE"))

mags_n_drep_ <- mags_n_drep %>% group_by(Group.2) %>% summarise(Total = sum(x))

mags_n_drep$x_ <- mags_n_drep$x
mags_n_drep$x_ <- ifelse(mags_n_drep$Group.2 != "OWC", mags_n_drep$x_, " ")
mags_n_drep$x_ <- ifelse(mags_n_drep$x > 1000, mags_n_drep$x_, " ")
mags_n_drep$x <- ifelse(mags_n_drep$Group.1 == "HQ", mags_n_drep$x + 1000, mags_n_drep$x)

plot_mags_n <-  ggplot() + 
                geom_bar(data=mags_n_drep, aes(fill=Group.1, y=x, x=Group.2), position="stack", stat="identity") + 
                geom_text(data = mags_n_drep_, aes(label = Total, y = Total, x= Group.2), size=1.75, position = position_stack(vjust = 1), color="black", hjust = -0.25) +
                geom_text(data=mags_n_drep, aes(label = x_, y = x, x=Group.2, fill=Group.1), size=1.5, position = position_stack(vjust = 0.5), color="white") +
                labs(title="", y="De-replicated MAGs", x="") + scale_y_continuous(expand=c(0,0), limits = c(0,95000)) + preset_gen + coord_flip()  +
                scale_fill_manual(name = "MAG\nquality:", values = c("HQ"="#1b9e77", "MQ" = "#d95f02")) + theme(legend.position = "right")
plot_mags_n
```


### Function for linking rRNA genes into clusters
```{r, include=TRUE}
rrna_proc <- function(genes,link_dist,category) {
    rrna <- read.delim(genes, sep="\t", header=T)
    rrna$group <- category
    rrna$link_c <- paste(rrna$bin,rrna$contig,sep="_")

    rrna <- rrna %>%
      arrange(link_c, start) %>%
      group_by(link_c) %>%
      mutate(
          prev_end = lag(end, default = 10**9),
          next_start = lead(start, default = 10**9),
          prev_seq = lag(rrna, default = "Contig_start"),
          next_seq = lead(rrna, default = "Contig_end"),
          clust_in = (abs(start - prev_end) <= link_dist) | (abs(next_start - end) <= link_dist),
          clust_end = !(abs(end - next_start) <= link_dist),
          clust_id = cumsum(lag(clust_end, default = TRUE))
      ) %>% ungroup()
    
    rrna$cluster <- paste0(rrna$link_c,"_cluster_",rrna$clust_id)

  return (rrna)}
```


### Function for checking rRNA cluster completeness
```{r, include=TRUE}
rrna_clust <- function(rrna) {
    # Gene counts
    rrna_n <- aggregate(rrna$cluster,by=list(rrna$bin,rrna$cluster), FUN=length)
    colnames(rrna_n) <- c("bin","cluster","rrna_n")
  
    # Cluster lengths
    rrna1 <- rrna[!duplicated(rrna$cluster),]
    rrna2 <- rrna[rev(seq_len(nrow(rrna))), ]
    rrna2 <- rrna2[!duplicated(rrna2$cluster),]
    rrna_len <- merge(rrna1[,c("cluster","start")],rrna2[,c("cluster","end")],by="cluster")
    rrna_len$len <- rrna_len$end - rrna_len$start
    
    # Content
    rrna_type <- rrna %>% group_by(cluster) %>% summarise(rrna_all = paste(unique(rrna), collapse = ", "))
    rrna_type$operon <- ifelse(grepl("16S",rrna_type$rrna_all) & grepl("5S",rrna_type$rrna_all) & grepl("23S",rrna_type$rrna_all), "Complete", "Incomplete")
  
    # Aggregate
    rrna_c <- merge(rrna_len,rrna_n,by="cluster")
    rrna_c <- merge(rrna_c,rrna_type[,c("cluster","rrna_all","operon")],by="cluster")
    
  return (rrna_c)}
```


### Link rRNA genes for catalogs
```{r, include=TRUE}
# Make clusters
link_dist <- 3000

rrna_lr <- rrna_proc("../data/catalogs/rrna/rrna_mfd-lr.tsv",link_dist,"MFD-LR")
rrna_lr <- rrna_lr[rrna_lr$bin %in% mfd_lr[mfd_lr$rep_status == "TRUE",]$bin,]

rrna_sr <- rrna_proc("../data/catalogs/rrna/rrna_mfd-sr.tsv",link_dist,"MFD-SR")
rrna_sr <- rrna_sr[rrna_sr$bin %in% mfd_sr_drep$bin,]

rrna_spire <- rrna_proc("../data/catalogs/rrna/rrna_spire.tsv",link_dist,"SPIRE")
rrna_smag <- rrna_proc("../data/catalogs/rrna/rrna_smag.tsv",link_dist,"SMAG")
rrna_rbg <- rrna_proc("../data/catalogs/rrna/rrna_rbg.tsv",link_dist,"RBG")
rrna_gem <- rrna_proc("../data/catalogs/rrna/rrna_gem.tsv",link_dist,"GEM")
rrna_tpmc <- rrna_proc("../data/catalogs/rrna/rrna_tpmc.tsv",link_dist,"TPMC")
rrna_owc <- rrna_proc("../data/catalogs/rrna/rrna_owc.tsv",link_dist,"OWC")

# Check completeness
rrna_lr_c <- rrna_clust(rrna_lr)
rrna_sr_c <- rrna_clust(rrna_sr)
rrna_spire_c <- rrna_clust(rrna_spire)
rrna_smag_c <- rrna_clust(rrna_smag)
rrna_rbg_c <- rrna_clust(rrna_rbg)
rrna_gem_c <- rrna_clust(rrna_gem)
rrna_tpmc_c <- rrna_clust(rrna_tpmc)
rrna_owc_c <- rrna_clust(rrna_owc)
```


### Plot rRNA cluster by catalog
```{r, include=TRUE}
rrna_c <- rbind(rrna_lr_c,rrna_sr_c,rrna_spire_c,rrna_rbg_c,rrna_gem_c,rrna_smag_c,rrna_tpmc_c,rrna_owc_c)
rrna <- rbind(rrna_lr,rrna_sr,rrna_spire,rrna_rbg,rrna_gem,rrna_smag,rrna_tpmc,rrna_owc)

rrna$group2 <- ifelse(grepl("16S",rrna$rrna), "16S rRNA",
                      ifelse(grepl("23S",rrna$rrna), "23S rRNA",
                             ifelse(grepl("5S",rrna$rrna), "5S rRNA", "Error")))

ssu <- aggregate(rrna$cluster,by=list(rrna$group,rrna$group2), FUN=length)
colnames(ssu) <- c("group","type","rrna_all")
ssu$link <- paste(ssu$group,ssu$type,sep="_")

ssu_compl <- aggregate(rrna[rrna$cluster %in% rrna_c[rrna_c$operon=="Complete",]$cluster,]$cluster,
                       by=list(rrna[rrna$cluster %in% rrna_c[rrna_c$operon=="Complete",]$cluster,]$group,
                               rrna[rrna$cluster %in% rrna_c[rrna_c$operon=="Complete",]$cluster,]$group2), FUN=length)

colnames(ssu_compl) <- c("group","type","rrna_operon")
ssu_compl$link <- paste(ssu_compl$group,ssu_compl$type,sep="_")

ssu <- merge(ssu,ssu_compl[,c("link","rrna_operon")],by="link")
ssu$frac <- round(ssu$rrna_operon/ssu$rrna_all*100,2)
ssu$frac2 <- paste0(format(round(ssu$frac,1),nsmall=1)," %")

ssu$group <- factor(ssu$group,levels = c("OWC","MFD-SR","TPMC","MFD-LR","GEM","SMAG","RBG","SPIRE"))

plot_rrna <- ggplot(data = ssu, mapping = aes(x = type, y = group, fill = frac)) + geom_tile(width=1) +
             geom_text(aes(label = rrna_all), size=1.75, vjust=-0.7) + geom_text(aes(label = frac2), size=1.75, vjust=1)  +
             labs(title="", y="", x="rRNA operons", fill="Estimated as complete (%):") + preset_gen2 
plot_rrna
```


### Load genome catalog IDs
```{r, include=TRUE}
drep_spire <- read.delim("../data/catalogs/id/mags_drep_spire.txt", sep="\t", header=F)
drep_rbg <- read.delim("../data/catalogs/id/mags_drep_rbg.txt", sep="\t", header=F)
drep_smag <- read.delim("../data/catalogs/id/mags_drep_smag.txt", sep="\t", header=F)
drep_gem <- read.delim("../data/catalogs/id/mags_drep_gem.txt", sep="\t", header=F)
drep_tpmc <- read.delim("../data/catalogs/id/mags_drep_tpmc.txt", sep="\t", header=F)
drep_owc <- read.delim("../data/catalogs/id/mags_drep_owc.txt", sep="\t", header=F)
```


### Function for wrangling Antismash results
```{r, include=TRUE}
as_proc <- function(load,category, sub) {
  # Load and wrangle
  as <- read.delim(load, sep="\t", header=T)
  colnames(as)[1] <- "bin"
  as <- as[as$bin %in% sub,]
  as$KCB_perc[is.na(as$KCB_perc)] <- 0
  as$group <- category
  
  # Classify
  PKS <- c("T1PKS", "T2PKS", "T3PKS", "HR-T2PKS", "PKS-like","transAT-PKS", "transAT-PKS-like", "hglE-KS", "prodigiosin")
  NRPS <- c("NRPS", "NRPS-like", "thioamide-NRP", "NAPAA")
  RiPPs <- c("lantipeptide","thiopeptide", "bacteriocin", "linaridin", "cyanobactin", "glycocin", "LAP", "lassopeptide","sactipeptide",
             "bottromycin", "microcin", "microviridin", "proteusin", "guanidinotides", "lanthipeptide", "lipolanthine", "RaS-RiPP", "fungal-RiPP",
             "thioamitides", "lanthipeptide-class-i", "lanthipeptide-class-ii", "lanthipeptide-class-iii", "lanthipeptide-class-iv",
             "lanthipeptide-class-v", "ranthipeptide", "redox-cofactor", "RRE-containing", "epipeptide", "cyclic-lactone-autoinducer", "spliceotide",
             "crocagin", "RiPP-like", "achaeal-RiPP")
  TERP <- c("terpene")
  SACH <- c("amglyccycl", "oligosaccharide", "cf_saccharide", "saccharide", "2dos")
  BL <- c("betalactone")
  METAL <- c("NI-siderophore", "opine-like-metallophore", "aminopolycarboxylic-acid")
  HCN <- c("hydrogen-cyanide")
  HL <- c("hserlactone")
  IND <- c("indole")
  ARL <- c("arylpolyene")
  
  as$class <- ifelse(as$product %in% PKS, "PKS", 
              ifelse(as$product %in% NRPS, "NRPS",
              ifelse(as$product %in% RiPPs,"RiPPs",
              ifelse(as$product %in% TERP,"Terpenes",
              ifelse(as$product %in% SACH,"Saccharides",
              ifelse(as$product %in% BL,"Betalactones",
              ifelse(as$product %in% HL,"Homoserine lactones",
              ifelse(as$product %in% METAL,"Metallophores",
              ifelse(as$product %in% HCN,"Hydrogen cyanides",
              ifelse(as$product %in% IND,"Indoles",
              ifelse(as$product %in% ARL,"Arylpolyenes",      
              ifelse(grepl("/",as$product),"Hybrids","Other"))))))))))))
  
  # Re-assign some hybrids
  as_all <- unique(as$product)
  as_single <- unique(as[!grepl("/",as$product),]$product)
  
  as$class <- ifelse(as$class == "Hybrids", 
              ifelse(grepl(paste(PKS,collapse="|"),as$product) & grepl(paste(NRPS,collapse="|"),as$product),
                     "PKS/NRPS hybrids","Hybrids"), as$class)

  as$class <- ifelse(as$class == "Hybrids",
              ifelse(grepl(paste(PKS,collapse="|"),as$product) & !grepl(paste(setdiff(as_single, PKS),collapse="|"),as$product),
                     "PKS","Hybrids"), as$class)

  as$class <- ifelse(as$class == "Hybrids",
              ifelse(grepl(paste(NRPS,collapse="|"),as$product) & !grepl(paste(setdiff(as_single, NRPS),collapse="|"),as$product),
                     "NRPS","Hybrids"), as$class)

  as$class <- ifelse(as$class == "Hybrids", 
              ifelse(grepl(paste(RiPPs,collapse="|"),as$product) & !grepl(paste(setdiff(as_single, RiPPs),collapse="|"),as$product),
                     "RiPPs","Hybrids"), as$class)
  return (as)}
```


### Load antismash results
```{r, include=TRUE}
as_lr <- as_proc("../data/catalogs/bgc/BGC_lr_sum.tsv","MFD-LR",mfd_lr[mfd_lr$rep_status == "TRUE",]$bin) 
as_sr <- as_proc("../data/catalogs/bgc/BGC_sr_sum.tsv","MFD-SR",mfd_sr_drep$bin) 
as_spire <- as_proc("../data/catalogs/bgc/BGC_spire_sum.tsv","SPIRE",drep_spire$V1) 
as_smag <- as_proc("../data/catalogs/bgc/BGC_smag_sum.tsv","SMAG",drep_smag$V1) 
as_rbg <- as_proc("../data/catalogs/bgc/BGC_rbg_sum.tsv","RBG",drep_rbg$V1) 
as_gem <- as_proc("../data/catalogs/bgc/BGC_gem_sum.tsv","GEM",drep_gem$V1) 
as_tpmc <- as_proc("../data/catalogs/bgc/BGC_tpmc_sum.tsv","TPMC",drep_tpmc$V1) 
as_owc <- as_proc("../data/catalogs/bgc/BGC_owc_sum.tsv","OWC",drep_owc$V1) 
```


### Plot BGCs by catalog
```{r, include=TRUE}
as <- rbind(as_lr,as_sr,as_smag,as_rbg,as_gem,as_tpmc,as_owc,as_spire)

as$link_t <- paste(as$group,as$class,sep="_")
as_sum <- aggregate(as$record_id,by=list(as$link_t), FUN=length)
colnames(as_sum) <- c("link","bgc_sum")

as_edge <- aggregate(as$record_id,by=list(as$link_t,as$group,as$class,as$contig_edge), FUN=length)
colnames(as_edge) <- c("link","group","type","edge","bgc")
as_edge <- merge(as_edge,as_sum, by="link")

as_edge$edge_frac <- round(as_edge$bgc/as_edge$bgc_sum*100,2)

as_filt <- as_edge[as_edge$type %in% c("PKS","NRPS","RiPPs","Terpenes","Betalactones", "Arylpolyenes"),]
as_filt <- as_filt[as_filt$edge == "False",]

as_filt$group <- factor(as_filt$group,levels = c("OWC","MFD-SR","TPMC","MFD-LR","GEM","SMAG","RBG","SPIRE"))
as_filt$type <- factor(as_filt$type,levels = c("PKS","NRPS","RiPPs","Terpenes","Betalactones", "Arylpolyenes"))

as_filt$edge_frac2 <- paste0(format(round(as_filt$edge_frac,1),nsmall=1)," %")

plot_bgc <- ggplot(data = as_filt, mapping = aes(x = type, y = group, fill = edge_frac)) + geom_tile(width=1) +
            geom_text(aes(label = bgc_sum), size=1.75, vjust=-0.7) +  geom_text(aes(label = edge_frac2), size=1.75, vjust=1) +
            labs(title="", y="", x="BGCs", fill="Estimated as complete (%):") + preset_gen2
plot_bgc
```


### Function for loading and wrangling defensome data
```{r, include=TRUE}
df_proc <- function(sys,genes,contigs,category) {
  
  df_genes <- read.delim(genes, sep="\t", header=T)
  df_genes$sub <- gsub("SoFic","SoFIC", df_genes$sub)
  
  df_genes_ <- read.delim(sys, sep="\t", header=T)
  df_genes_$sys <- gsub("Cas","CRISPR-Cas", df_genes_$sys)
  df_genes_$sub <- gsub("SoFic","SoFIC", df_genes_$sub)
  df_genes_ <- df_genes_ %>% select(sys,sub) %>% distinct()
  
  df_genes <- merge(df_genes,df_genes_,by="sub")
  df_genes$link <- paste(df_genes$bin,df_genes$sys,sep="_")
  df_genes$link_c <- paste(df_genes$bin,df_genes$contig,sep="_")
  
  con_len <- read.delim(contigs, sep="\t", header=T)
  con_len$contig <- sub(" .*", "", con_len$contig)
  con_len$link_c <- paste(con_len$bin,con_len$contig,sep="_")
  df_genes <- merge(df_genes,con_len[,c("link_c","len")],by="link_c")
  df_genes$group <- category

  return (df_genes)}
```


### Get defensome data for catalogs
```{r, include=TRUE}
df_lr <- df_proc("../data/catalogs/df/df_mfd-lr_sys.tsv","../data/catalogs/df/df_mfd-lr_bakta.tsv","../data/catalogs/df/contigs_mfd-lr_sub.tsv","MFD-LR")
df_lr <- df_lr[df_lr$bin %in% mfd_lr[mfd_lr$rep_status == "TRUE",]$bin,]

df_sr <- df_proc("../data/catalogs/df/df_mfd-sr_sys.tsv","../data/catalogs/df/df_mfd-sr_bakta.tsv","../data/catalogs/df/contigs_mfd-sr_sub.tsv","MFD-SR")
df_sr <- df_sr[df_sr$bin %in% mfd_sr_drep$bin,]

df_spire <- df_proc("../data/catalogs/df/df_spire_sys.tsv","../data/catalogs/df/df_spire_bakta.tsv","../data/catalogs/df/contigs_spire_sub.tsv","SPIRE")
df_rbg <- df_proc("../data/catalogs/df/df_rbg_sys.tsv","../data/catalogs/df/df_rbg_bakta.tsv","../data/catalogs/df/contigs_rbg_sub.tsv","RBG")
df_smag <- df_proc("../data/catalogs/df/df_smag_sys.tsv","../data/catalogs/df/df_smag_bakta.tsv","../data/catalogs/df/contigs_smag_sub.tsv","SMAG")
df_gem <- df_proc("../data/catalogs/df/df_gem_sys.tsv","../data/catalogs/df/df_gem_bakta.tsv","../data/catalogs/df/contigs_gem_sub.tsv","GEM")
df_tpmc <- df_proc("../data/catalogs/df/df_tpmc_sys.tsv","../data/catalogs/df/df_tpmc_bakta.tsv","../data/catalogs/df/contigs_tpmc_sub.tsv","TPMC")
df_owc <- df_proc("../data/catalogs/df/df_owc_sys.tsv","../data/catalogs/df/df_owc_bakta.tsv","../data/catalogs/df/contigs_owc_sub.tsv","OWC")
```


### Function for clustering defense genes
```{r, include=TRUE}
df_clust <- function(df,link_dist,edge_dist) {

    df <- df %>%
      arrange(link_c, start) %>%
      group_by(link_c) %>%
      mutate(
          prev_end = lag(end, default = 10**9),
          next_start = lead(start, default = 10**9),
          prev_sys = lag(sys, default = "Contig_start"),
          next_sys = lead(sys, default = "Contig_end"),
          clust_in = (abs(start - prev_end) <= link_dist) | (abs(next_start - end) <= link_dist),
          clust_end = !(abs(end - next_start) <= link_dist),
          clust_id = cumsum(lag(clust_end, default = TRUE))
      ) %>% ungroup()

    df$cluster <- paste0(df$link_c,"_cluster_",df$clust_id)
    
    df1 <- df[!duplicated(df$cluster),]
    df1 <- df1[df1$start <= edge_dist,]
    
    df2 <- df[rev(seq_len(nrow(df))), ]
    df2 <- df2[!duplicated(df2$cluster),]
    df2 <- df2[(df2$len - df2$end) <= edge_dist,]
    
    df$clust_edge <- ifelse(df$cluster %in% df1$cluster, TRUE, 
                     ifelse(df$cluster %in% df2$cluster, TRUE, FALSE))
    
  return (df)}
```


### Function for classifying defense gene clusters
```{r, include=TRUE}
df_sum <- function(df,gene_min,fam_min) {
  # Gene counts
  df_n <- aggregate(df$cluster,by=list(df$bin,df$cluster), FUN=length)
  colnames(df_n) <- c("bin","cluster","df_genes")
  
  # System counts
  df$link_f <- paste(df$cluster,df$sys,sep="_")
  df_ <- df[!duplicated(df$link_f),]
  df_fam <- aggregate(df_$cluster,by=list(df_$cluster), FUN=length)
  colnames(df_fam) <- c("cluster","df_sys")
  df_n <- merge(df_n,df_fam, by="cluster")
  
  # Cluster lengths
  df1 <- df[!duplicated(df$cluster),]
  df2 <- df[rev(seq_len(nrow(df))), ]
  df2 <- df2[!duplicated(df2$cluster),]
  df_len <- merge(df1[,c("cluster","start")],df2[,c("cluster","end")],by="cluster")
  df_len$len <- df_len$end - df_len$start
  df_n <- merge(df_n,df_len,by="cluster")
  
  # Fragmentation and island status
  df_n$clust_edge <- ifelse(df_n$cluster %in% df[df$clust_edge == FALSE,]$cluster, FALSE, TRUE)
  df_n$island <- ifelse(df_n$df_genes >= gene_min & df_n$df_sys >= fam_min, TRUE, FALSE)
  df_n$group <- unique(df$group)
  
  # System content
  df_sys <- df %>% group_by(cluster) %>% summarise(sys_all = paste(unique(sys), collapse = ", "))
  df_n <- merge(df_n,df_sys[,c("cluster","sys_all")],by="cluster")
  df_n$type <- ifelse(df_n$df_sys == 1, df_n$sys_all, "Hybrid")
  
  return (df_n)}
```


### Process defense genes for catalogs
```{r, include=TRUE}
# Cluster
link_dist <- 20000
edge_dist <- 1000

df_lr_ <- df_clust(df_lr,link_dist,edge_dist)
df_sr_ <- df_clust(df_sr,link_dist,edge_dist)
df_spire_ <- df_clust(df_spire,link_dist,edge_dist)
df_rbg_ <- df_clust(df_rbg,link_dist,edge_dist)
df_smag_ <- df_clust(df_smag,link_dist,edge_dist)
df_gem_ <- df_clust(df_gem,link_dist,edge_dist)
df_tpmc_ <- df_clust(df_tpmc,link_dist,edge_dist)
df_owc_ <- df_clust(df_owc,link_dist,edge_dist)

# Summarise
gene_min <- 3
fam_min <- 0

df_lr_c <- df_sum(df_lr_,gene_min,fam_min)
df_sr_c <- df_sum(df_sr_,gene_min,fam_min)
df_spire_c <- df_sum(df_spire_,gene_min,fam_min)
df_rbg_c <- df_sum(df_rbg_,gene_min,fam_min)
df_smag_c <- df_sum(df_smag_,gene_min,fam_min)
df_gem_c <- df_sum(df_gem_,gene_min,fam_min)
df_tpmc_c <- df_sum(df_tpmc_,gene_min,fam_min)
df_owc_c <- df_sum(df_owc_,gene_min,fam_min)

df_c <- rbind(df_lr_c,df_sr_c,df_spire_c,df_rbg_c,df_smag_c,df_gem_c,df_tpmc_c,df_owc_c)
```


### Plot defense island count summary
```{r, include=TRUE}
# Get fragmentation counts
df_c_ <- df_c[df_c$island == TRUE,]

df_c_agr_all <- aggregate(df_c_$cluster,by=list(df_c_$type,df_c_$group), FUN=length)
colnames(df_c_agr_all) <- c("type","group","di_all")
df_c_agr_all$link <- paste0(df_c_agr_all$group,"_",df_c_agr_all$type)

df_c_agr_full <- aggregate(df_c_[df_c_$clust_edge == FALSE,]$cluster,by=list(df_c_[df_c_$clust_edge == FALSE,]$type,df_c_[df_c_$clust_edge == FALSE,]$group), FUN=length)
colnames(df_c_agr_full) <- c("type","group","di_full")
df_c_agr_full$link <- paste0(df_c_agr_full$group,"_",df_c_agr_full$type)

df_c_agr <- merge(df_c_agr_all,df_c_agr_full[,c("link","di_full")],by="link", all=T)
df_c_agr[is.na(df_c_agr)] <- 0

df_c_agr$frac <- round(df_c_agr$di_full/df_c_agr$di_all*100,2)

# Subset to most common islands and order
df_c_n <- aggregate(df_c_$cluster,by=list(df_c_$type), FUN=length)
colnames(df_c_n) <- c("type","type_sum_all")

df_c_n <- df_c_n[order(df_c_n$type_sum_all,decreasing=TRUE),]
df_c_agr <- df_c_agr[df_c_agr$type %in% head(df_c_n$type,5),]

df_c_agr$group <- factor(df_c_agr$group,levels = c("OWC","MFD-SR","TPMC","MFD-LR","GEM","SMAG","RBG","SPIRE"))
df_c_agr$type <- factor(df_c_agr$type,levels = head(df_c_n$type,5))

df_c_agr$frac2 <- paste0(format(round(df_c_agr$frac,1),nsmall=1)," %")

df_c_agr <- df_c_agr[df_c_agr$type != "Hybrid",]

plot_df <- ggplot(data = df_c_agr, mapping = aes(x = type, y = group, fill = frac)) + geom_tile(width=1) +
           geom_text(aes(label = di_all), size=1.75, vjust=-0.7) + geom_text(aes(label = frac2), size=1.75, vjust=1) + 
           labs(title="", y="", x="Defense islands", fill="Estimated as complete (%):") + preset_gen2
plot_df
```


### Plot Figure 3
```{r, include=TRUE}
# First half
plot_table_ <- plot_table + theme(plot.margin = unit(c(-0.2,0,0,0.25), 'cm'))
plot_upset_ <- as.ggplot(plot_upset) + theme(plot.margin = unit(c(-0.15,-1.8,-0.65,-0.25), 'cm'))

plot_mags_n_ <- plot_mags_n+ guides(fill = guide_legend(override.aes = list(size=1),reverse=T)) +
                theme(legend.key.size = unit(0.3, "cm"),legend.margin = unit(c(-3,0,0,0), 'cm'),
                      legend.box.background = element_blank(),legend.background = element_blank())

plot_upset_ <- ggarrange(plot_upset_, as_ggplot(get_legend(plot_mags_n_)), nrow=1, ncol=2, widths = c(1,0.2), align = "none")

fig_ab <- ggarrange(plot_table_,plot_upset_, nrow=1, ncol=2, widths = c(1,1.5), common.legend = FALSE,
                    align = c("v"), labels = c("a","b"), font.label = list(size = 9))

# Second half
margins <- list(theme(legend.key.width = unit(0.4, "cm"),  legend.key.height = unit(0.35, "cm"), 
                      legend.box.spacing = unit(0, "cm"), legend.box.margin = unit(c(0.1,0,-0.2,-0.7), 'cm'),
                      legend.box.background = element_blank(), legend.key = element_blank(), 
                      legend.background = element_blank(), plot.margin = unit(c(-0.2,0.1,-0.6,-0.3), 'cm')))

plot_df_ <- plot_df + margins + theme(axis.text.y=element_blank(), plot.margin = unit(c(-0.2,0.1,-0.6,-0.15), 'cm'))
plot_bgc_ <- plot_bgc + margins + theme(axis.text.y=element_blank(), plot.margin = unit(c(-0.2,0.1,-0.6,-0.15), 'cm'))
plot_rrna_ <- plot_rrna + margins

fig_ce <- ggarrange(plot_rrna_,plot_df_,plot_bgc_, nrow=1, ncol=3, widths = c(1,1,1.35), common.legend = TRUE, legend="bottom", 
                    align = c("h"), labels = c("c","d", "e"), font.label = list(size = 9))

plot_sylph_sr <- readRDS("../data/catalogs/sylph.rds") +
                 theme(plot.margin = unit(c(-0.2,0.1,0,0.1), 'cm'),
                       legend.box.spacing = unit(-0.1, "cm"), legend.background = element_blank(),
                       legend.box.background = element_blank(), legend.key.size = unit(0.1, "cm"),
                       legend.key.height = unit(0.01, 'cm'), legend.key.width = unit(0.1, 'cm'),
                       legend.key = element_rect(color = NA, fill = NA), legend.spacing.y = unit(-0.5, "lines"),
                       legend.text = element_text(lineheight = 0.01, vjust = 0)) +
                 guides(color = guide_legend(ncol = 1, byrow = TRUE, override.aes = list(size = 1.5))) 

fig_cf <- ggarrange(fig_ce, plot_sylph_sr, nrow=1, ncol=2, widths = c(1.5,1), common.legend = FALSE, legend="bottom",
                    align = c("none"), labels = c("", "f"), font.label = list(size = 9))

# Merge
fig_full <- ggarrange(fig_ab,fig_cf, nrow=2, ncol=1, heights = c(1,1.1), common.legend = FALSE, align = c("hv"))
fig_full

#ggsave(file="../analysis/figures/Fig3.pdf", height = 140, width = 180, useDingbats=FALSE,  units ="mm", dpi=300)
```

