---
title: "R code for plotting data related to the Microflora Danica ~10,000 shallow metagenomes"
author: "Mantas Sereika"
date: "2025"
output:
  html_document: default
  pdf_document: default
---


### Load dependencies
```{r, include=TRUE, message=FALSE}
library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggbeeswarm)
library(ggpubr)
```


### Load 16S-related data
```{r, include=TRUE}
mags <- read.delim("../analysis/datasets/dataset_S3.tsv", sep="\t", header=T)
mags_16s_lr <- read.delim("../data/MFD-SR/MFG_MFD-LR.tsv", sep="\t", header=F)
mags_16s_sr <- read.delim("../data/MFD-SR/MFG_MFD-SR.tsv", sep="\t", header=F)

core_type <- read.delim("../data/MFD-SR/2024-03-07_core-genera-type.csv", sep=",", header=T)
core_area <- read.delim("../data/MFD-SR/2024-03-07_core-genera-area.csv", sep=",", header=T)
core_hab1 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo1.csv", sep=",", header=T)
core_hab2 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo2.csv", sep=",", header=T)
core_hab3 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo3.csv", sep=",", header=T)
```


### Plot presets
```{r, include=TRUE}
preset_gen <- list(theme_bw(),
                   theme(legend.position = "bottom",
                         legend.text=element_text(size=7),
                         legend.title = element_text(size=7),
                         axis.title.y = element_text(size = 7), 
                         axis.title.x = element_text(size = 7),
                         axis.text.y = element_text(size = 6),
                         axis.text.x = element_text(size = 6),
                         panel.grid.minor = element_line(size = 0.15),
                         panel.grid.major = element_line(size = 0.3)))

preset_gen2 <- list(preset_gen, coord_flip(),
                    theme(legend.position = "none",axis.ticks = element_blank()))
```


### Wrangle MAG 16S taxonomy data
```{r, include=TRUE}
# MFD-LR
mags_16s_lr$bin <- sapply(strsplit(mags_16s_lr$V1, "_"), `[`, 1)

mags_16s_lr <- separate(data = mags_16s_lr, col = V2, sep = ",", into = c("ASV_domain", "phylum","class","order","family","genus","species"))
mags_16s_lr <- separate(data = mags_16s_lr, col = ASV_domain, sep = ";tax=d:",into = c("ASV", "domain"))

mags_16s_lr$genus <- gsub("g:","",mags_16s_lr$genus)
mags_16s_lr$genus <- ifelse(mags_16s_lr$V3 >= 94.5, mags_16s_lr$genus, "Novel genus")

mags_16s_lr_hq <- mags_16s_lr[mags_16s_lr$bin %in% mags[mags$MAG_status == "HQ",]$bin,]

# MFD-SR
mags_16s_sr$bin <- sapply(strsplit(mags_16s_sr$V1, "_16S_"), `[`, 1)
mags_16s_sr$genus <- gsub("g:","",sapply(strsplit(mags_16s_sr$V2, ","), `[`, 6))
mags_16s_sr$genus <- ifelse(mags_16s_sr$V3 >= 94.5, mags_16s_sr$genus, "Novel genus")
```


### Core genera coverage by MAGs
```{r, include=TRUE}
core_cov <- function(df,mags_hq,mags_mq,mags_sr) {
  df$coverage <- ifelse(df$Genus %in% mags_hq$genus, "HQ MAG",
                        ifelse(df$Genus %in% mags_mq$genus, "MQ MAG",
                               ifelse(df$Genus %in% mags_sr$genus,"SR MAG","No MAG")))
  return (df) }

core_type <- core_cov(core_type,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_area <- core_cov(core_area,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab1 <- core_cov(core_hab1,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab2 <- core_cov(core_hab2,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab3 <- core_cov(core_hab3,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
```


### Overall coverage of core genera
```{r, include=TRUE}
core_agr <- function(df,type) {
  df_uniq <- df[!duplicated(df$Genus),]
  df_uniq_agr <- aggregate(df_uniq$Genus, by=list(df_uniq$coverage), FUN=length)
  colnames(df_uniq_agr) <- c("coverage","genera") 
  df_uniq_agr$group <- type
  return (df_uniq_agr) }

agr_type <- core_agr(core_type,"Sample type")
agr_area <- core_agr(core_area,"Area type")
agr_hab1 <- core_agr(core_hab1,"Habitat (level 1)")
agr_hab2 <- core_agr(core_hab2,"Habitat (level 2)")
agr_hab3 <- core_agr(core_hab3,"Habitat (level 3)")

agr_core <- rbind(agr_type,agr_area,agr_hab1,agr_hab2,agr_hab3)

agr_core_sum <- aggregate(agr_core$genera, by=list(agr_core$group), FUN=sum)
colnames(agr_core_sum) <- c("group","group_total")
agr_core <- merge(agr_core,agr_core_sum,by="group")
agr_core$genera_perc <- round(agr_core$genera/agr_core$group_total*100,2)

agr_core$coverage <- factor(agr_core$coverage,levels = rev(c("HQ MAG","MQ MAG","SR MAG","No MAG")))

agr_core$group <- factor(agr_core$group, levels = c("Sample type","Area type","Habitat (level 1)", "Habitat (level 2)", "Habitat (level 3)"))
```


### Plot overall coverage by all MFD MAGs
```{r, include=TRUE}
plot_cov_all <- ggplot(agr_core[! agr_core$coverage == "No MAG",], aes(y=genera_perc, x=group, fill=coverage)) + 
                geom_bar(position="stack", stat="identity") + preset_gen + 
                geom_text(aes(label = genera), size=2, position=position_stack(vjust = 0.5)) +
                scale_y_continuous(breaks=c(0,20,40,60,80,100), expand=c(0,0), limits = c(0,105)) +
                labs(x="", y="Coverage of core genera (%)", fill="Coverage type:") +
                scale_fill_manual(values=c("HQ MAG"="#1b9e77", "MQ MAG"="#d95f02", "SR MAG"="#fee090"), guide = guide_legend(reverse = TRUE)) 
plot_cov_all
```


### Plot overall coverage by long-read MAGs
```{r, include=TRUE}
plot_cov_lr <- ggplot(agr_core[! (agr_core$coverage == "No MAG" | agr_core$coverage == "SR MAG"),], aes(y=genera_perc, x=group, fill=coverage)) + 
               geom_bar(position="stack", stat="identity") + preset_gen + 
               geom_text(aes(label = genera), size=2, position=position_stack(vjust = 0.5)) +
               scale_y_continuous(breaks=c(0,20,40,60,80,100), expand=c(0,0), limits = c(0,105)) +
               labs(x="", y="Coverage of core genera (%)", fill="Coverage type:") +
               scale_fill_manual(values=c("HQ MAG"="#1b9e77", "MQ MAG"="#d95f02"), guide = guide_legend(reverse = TRUE)) 
plot_cov_lr
#saveRDS(plot_cov_lr,"../data/MFD-LR/core_genera.rds")
```


### Calculate coverage ranks per habitat
```{r, include=TRUE}
core_agr_hab <- function(df,type) {
  df$coverage2 <- ifelse(df$coverage == "No MAG" | df$coverage == "SR MAG",0,1)
  df_sum1 <- aggregate(df$coverage2, by=list(df[,6]), FUN=sum)
  colnames(df_sum1) <- c("complex","genera_mag")
  df_sum2 <- aggregate(df$Genus, by=list(df[,6]), FUN=length)
  colnames(df_sum2) <- c("complex","genera_total")
  df_sum <- merge(df_sum1,df_sum2,by="complex")
  df_sum$genera_perc <- round(df_sum$genera_mag/df_sum$genera_total*100,2)
  df_sum$complex0 <- df_sum$complex
  df_sum <- separate(data=df_sum, col=complex0, sep=",", extra="merge",into=c("type", "area", "hab"))
  df_sum$group <- type
  return (df_sum) }

core_area_sum <- core_agr_hab(core_area,"Sample area")
core_hab1_sum <- core_agr_hab(core_hab1,"Habitat (level 1)")
core_hab2_sum <- core_agr_hab(core_hab2,"Habitat (level 2)")
core_hab3_sum <- core_agr_hab(core_hab3,"Habitat (level 3)")

core_hab_sum <- rbind(core_area_sum,core_hab1_sum,core_hab2_sum,core_hab3_sum)
core_hab_sum$type <- factor(core_hab_sum$type,levels =c("Soil","Sediment","Water"))
core_hab_sum$area <- factor(core_hab_sum$area,levels =c(" Agriculture"," Natural"," Urban", " Subterranean"))
```


### Plot core genera coverage at different metadata levels
```{r, include=TRUE}
core_hab_sum <- core_hab_sum[order(core_hab_sum$type,decreasing=TRUE),]

preset <- list(preset_gen2, geom_quasirandom(size = 0.6, width = 0.2),
               geom_violin(scale="width", width=0.25, lwd=0.25,linetype="dashed", fill=NA, alpha=0.5),
               scale_y_continuous(expand=c(0,0), limits = c(0,102),breaks=c(0,20,40,60,80,100)),
               scale_color_manual(values=c("Soil"="#8c510a", "Sediment"="#4d9221", "Water"="#3288bd")))

plot_core_area <- ggplot(data=core_hab_sum[core_hab_sum$group=="Sample area",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for sample area category (%)", x="") + preset

plot_core_hab1 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 1)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 1 (%)", x="") + preset

plot_core_hab2 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 2)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 2 (%)", x="") + preset

plot_core_hab3 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 3)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 3 (%)", x="") + preset 

plots_core_hab <- ggarrange(plot_core_area, plot_core_hab1, plot_core_hab2, plot_core_hab3, nrow=2, ncol=2, align = c("v"),
                            legend = "none", common.legend = FALSE, labels = c("a","b","c","d"), font.label = list(size = 9))

#ggsave(plots_core_hab, file="../analysis/figures/FigS8.pdf", height = 90, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Load sylph-related data
```{r, include=TRUE}
meta <- read.delim("../data/MFD-SR/2023-12-21_mfd_db.csv", sep=",", header=T)
meta <-  meta[, c("fieldsample_barcode","habitat_typenumber","mfd_sampletype","mfd_areatype", "mfd_hab1","mfd_hab2","mfd_hab3", "longitude", "latitude")]
meta$mfd_hab1 <- ifelse(meta$mfd_hab1 == "", "Other", meta$mfd_hab1)

mfd_lib <- read.delim("../data/MFD-SR/MFD_LIB.tsv", sep="\t", header=F)
colnames(mfd_lib) <- c("fieldsample_barcode","lib_id")

sylph_gtdb <- read.delim("../data/MFD-SR/sylph-gtdb.tsv", sep="\t", header=F)
sylph_gtdb$group <- "GTDB"

sylph_gtdb_pub <- read.delim("../data/MFD-SR/sylph-gtdb-pub.tsv", sep="\t", header=F)
sylph_gtdb_pub$group <- "GTDB + Public catalogues"

sylph_gtdb_pub_lr <- read.delim("../data/MFD-SR/sylph-gtdb-pub-lr.tsv", sep="\t", header=F)
sylph_gtdb_pub_lr$group <- "GTDB + Public catalogues + MFD-LR"

sylph_sr <- rbind(sylph_gtdb,sylph_gtdb_pub,sylph_gtdb_pub_lr)
colnames(sylph_sr) <- c("lib_id","class","group")

sylph_sr <- merge(sylph_sr,mfd_lib,by="lib_id")
sylph_sr <- merge(sylph_sr,meta,by="fieldsample_barcode")
```


### Plot classification rates by sample type
```{r, include=TRUE}
sylph_sr$group <- factor(sylph_sr$group,levels = c("GTDB","GTDB + Public catalogues","GTDB + Public catalogues + MFD-LR"))

plot_sylph_sr <- ggplot(data=sylph_sr, aes(x=mfd_sampletype, y=class, by=group, color=group)) + 
                 geom_quasirandom(size = 0.01, alpha=0.4, dodge.width = 0.9) + 
                 geom_boxplot(outlier.shape=NA, fill=NA, width=0.3, alpha = 0.4, lwd=0.1, position = position_dodge(width = 0.9), color="black") + 
                 labs(title="", y="Short reads classified (%)", x="Sample type", color="Database:") + 
                 scale_x_discrete(guide = guide_axis(angle = 0)) + scale_y_continuous(expand=c(0,0), breaks = c(0,25,50,75,100), limits=c(0,102)) + 
                 scale_color_manual(values = c("#e7298a","#66a61e","#e6ab02")) +  preset_gen + 
                 guides(colour = guide_legend(override.aes = list(size=1.5), nrow=3, byrow=TRUE)) +
                 theme(legend.justification = "center", legend.spacing.x = unit(0.1, 'cm'), legend.spacing.y = unit(0.1, 'cm'), 
                       legend.key.height=unit(0.1, 'cm'), legend.key.width=unit(0.1, 'cm'), legend.box.spacing = unit(0, "cm"))

#saveRDS(plot_sylph_sr,"../data/catalogs/sylph.rds")
```


### Plot classification rates by sample type and habitat
```{r, include=TRUE}
preset_sylph <- list(labs(title="", y="Short reads classified (%)", x="", color="Database:"),
                     geom_quasirandom(size = 0.01, alpha=0.4, dodge.width = 0.9),
                     geom_boxplot(outlier.shape=NA, fill=NA, width=0.3, alpha = 0.4, lwd=0.1, position = position_dodge(width = 0.9), color="black"), 
                     scale_x_discrete(guide = guide_axis(angle = 45)),
                     scale_y_continuous(expand=c(0,0), breaks = c(0,25,50,75,100), limits=c(0,102)),
                     scale_color_manual(values = c("#e7298a","#66a61e","#e6ab02")),
                     guides(colour = guide_legend(override.aes = list(size=1.5), nrow=1, byrow=TRUE)),
                     facet_grid(cols = vars(mfd_areatype), scales = "free", space = "free"),
                     preset_gen, theme(strip.text = element_text(size=6)))

plot_sylph_sr_soil <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Soil",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                      labs(x="Sample habitat (soil sample type)") + preset_sylph + theme(legend.position = "none")

plot_sylph_sr_sed <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Sediment",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                     labs(x="Sample habitat (sediment sample type)") + preset_sylph

plot_sylph_sr_water <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Water",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                       labs(x="Sample habitat (water sample type)") + preset_sylph

fig_sylph_hab_ <- ggarrange(plot_sylph_sr_sed,plot_sylph_sr_water, nrow=1, ncol=2, common.legend = TRUE, legend="bottom", 
                            align = c("v"), labels = c("b","c"), font.label = list(size = 9))

fig_sylph_hab <- ggarrange(plot_sylph_sr_soil,fig_sylph_hab_, nrow=2, ncol=1, common.legend = FALSE, align = c("v"), 
                           labels = c("a",""), font.label = list(size = 9))

#ggsave(fig_sylph_hab, file="../analysis/figures/FigS9.pdf", height = 180, width = 200, useDingbats=FALSE,  units ="mm", dpi = 300)
```

