---
title: "R code for plotting data related to the Microflora Danica ~10,000 shallow metagenomes"
author: "Mantas Sereika"
date: "2025"
output:
  html_document: default
  pdf_document: default
---


### Load dependencies
```{r, include=TRUE, message=FALSE}
library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggbeeswarm)
library(ggpubr)
library(mapDK)
library(ape)
library(vegan) 
```


### Load 16S-related data
```{r, include=TRUE}
mags <- read.delim("../analysis/datasets/dataset_S3.tsv", sep="\t", header=T)
mags_16s_lr <- read.delim("../data/MFD-SR/MFG_MFD-LR.tsv", sep="\t", header=F)
mags_16s_sr <- read.delim("../data/MFD-SR/MFG_MFD-SR.tsv", sep="\t", header=F)

core_type <- read.delim("../data/MFD-SR/2024-03-07_core-genera-type.csv", sep=",", header=T)
core_area <- read.delim("../data/MFD-SR/2024-03-07_core-genera-area.csv", sep=",", header=T)
core_hab1 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo1.csv", sep=",", header=T)
core_hab2 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo2.csv", sep=",", header=T)
core_hab3 <- read.delim("../data/MFD-SR/2024-03-07_core-genera-mfdo3.csv", sep=",", header=T)
```


### Plot presets
```{r, include=TRUE}
preset_gen <- list(theme_bw(),
                   theme(legend.position = "bottom",
                         legend.text=element_text(size=7),
                         legend.title = element_text(size=7),
                         axis.title.y = element_text(size = 7), 
                         axis.title.x = element_text(size = 7),
                         axis.text.y = element_text(size = 6),
                         axis.text.x = element_text(size = 6),
                         panel.grid.minor = element_line(size = 0.15),
                         panel.grid.major = element_line(size = 0.3)))

preset_gen2 <- list(preset_gen, coord_flip(),
                    theme(legend.position = "none",axis.ticks = element_blank()))
```


### Function for getting medians and IQR
```{r, include=TRUE}
median_w_iqr <- function(x,cat) {
  df_med <- aggregate(x, by=list(cat), FUN=median, na.rm = T)
  colnames(df_med) <- c("group","median")
  
  df_iqr <- aggregate(x, by=list(cat), FUN=IQR, na.rm = T)
  colnames(df_iqr) <- c("group","iqr")
  
  df_q25 <- aggregate(x, by=list(cat), FUN = function(i,...) quantile(i, probs=0.25, na.rm=T))
  colnames(df_q25) <- c("group","q25")
  
  df_q75 <- aggregate(x, by=list(cat), FUN = function(i,...) quantile(i, probs=0.75, na.rm=T))
  colnames(df_q75) <- c("group","q75")
  
  df <- merge(df_med,df_iqr,by="group")
  df <- merge(df,df_q25,by="group")
  df <- merge(df,df_q75,by="group")
  return (df) }
```


### Wrangle MAG 16S taxonomy data
```{r, include=TRUE}
# MFD-LR
mags_16s_lr$bin <- sapply(strsplit(mags_16s_lr$V1, "_"), `[`, 1)

mags_16s_lr <- separate(data = mags_16s_lr, col = V2, sep = ",", into = c("ASV_domain", "phylum","class","order","family","genus","species"))
mags_16s_lr <- separate(data = mags_16s_lr, col = ASV_domain, sep = ";tax=d:",into = c("ASV", "domain"))

mags_16s_lr$genus <- gsub("g:","",mags_16s_lr$genus)
mags_16s_lr$genus <- ifelse(mags_16s_lr$V3 >= 94.5, mags_16s_lr$genus, "Novel genus")

mags_16s_lr_hq <- mags_16s_lr[mags_16s_lr$bin %in% mags[mags$MAG_status == "HQ",]$bin,]

# MFD-SR
mags_16s_sr$bin <- sapply(strsplit(mags_16s_sr$V1, "_16S_"), `[`, 1)
mags_16s_sr$genus <- gsub("g:","",sapply(strsplit(mags_16s_sr$V2, ","), `[`, 6))
mags_16s_sr$genus <- ifelse(mags_16s_sr$V3 >= 94.5, mags_16s_sr$genus, "Novel genus")
```


### Core genera coverage by MAGs
```{r, include=TRUE}
core_cov <- function(df,mags_hq,mags_mq,mags_sr) {
  df$coverage <- ifelse(df$Genus %in% mags_hq$genus, "HQ MAG",
                        ifelse(df$Genus %in% mags_mq$genus, "MQ MAG",
                               ifelse(df$Genus %in% mags_sr$genus,"SR MAG","No MAG")))
  return (df) }

core_type <- core_cov(core_type,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_area <- core_cov(core_area,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab1 <- core_cov(core_hab1,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab2 <- core_cov(core_hab2,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
core_hab3 <- core_cov(core_hab3,mags_16s_lr_hq,mags_16s_lr,mags_16s_sr)
```


### Overall coverage of core genera
```{r, include=TRUE}
core_agr <- function(df,type) {
  df_uniq <- df[!duplicated(df$Genus),]
  df_uniq_agr <- aggregate(df_uniq$Genus, by=list(df_uniq$coverage), FUN=length)
  colnames(df_uniq_agr) <- c("coverage","genera") 
  df_uniq_agr$group <- type
  return (df_uniq_agr) }

agr_type <- core_agr(core_type,"Sample type")
agr_area <- core_agr(core_area,"Area type")
agr_hab1 <- core_agr(core_hab1,"Habitat (level 1)")
agr_hab2 <- core_agr(core_hab2,"Habitat (level 2)")
agr_hab3 <- core_agr(core_hab3,"Habitat (level 3)")

agr_core <- rbind(agr_type,agr_area,agr_hab1,agr_hab2,agr_hab3)

agr_core_sum <- aggregate(agr_core$genera, by=list(agr_core$group), FUN=sum)
colnames(agr_core_sum) <- c("group","group_total")
agr_core <- merge(agr_core,agr_core_sum,by="group")
agr_core$genera_perc <- round(agr_core$genera/agr_core$group_total*100,2)

agr_core$coverage <- factor(agr_core$coverage,levels = rev(c("HQ MAG","MQ MAG","SR MAG","No MAG")))

agr_core$group <- factor(agr_core$group, levels = c("Sample type","Area type","Habitat (level 1)", "Habitat (level 2)", "Habitat (level 3)"))
```


### Plot overall coverage by all MFD MAGs
```{r, include=TRUE}
plot_cov_all <- ggplot(agr_core[! agr_core$coverage == "No MAG",], aes(y=genera_perc, x=group, fill=coverage)) + 
                geom_bar(position="stack", stat="identity") + preset_gen + 
                geom_text(aes(label = genera), size=2, position=position_stack(vjust = 0.5)) +
                scale_y_continuous(breaks=c(0,20,40,60,80,100), expand=c(0,0), limits = c(0,105)) +
                labs(x="", y="Coverage of core genera (%)", fill="Coverage type:") +
                scale_fill_manual(values=c("HQ MAG"="#1b9e77", "MQ MAG"="#d95f02", "SR MAG"="#fee090"), guide = guide_legend(reverse = TRUE)) 
plot_cov_all
```


### Plot overall coverage by long-read MAGs
```{r, include=TRUE}
plot_cov_lr <- ggplot(agr_core[! (agr_core$coverage == "No MAG" | agr_core$coverage == "SR MAG"),], aes(y=genera_perc, x=group, fill=coverage)) + 
               geom_bar(position="stack", stat="identity") + preset_gen + 
               geom_text(aes(label = genera), size=2, position=position_stack(vjust = 0.5)) +
               scale_y_continuous(breaks=c(0,20,40,60,80,100), expand=c(0,0), limits = c(0,105)) +
               labs(x="", y="Coverage of core genera (%)", fill="Coverage type:") +
               scale_fill_manual(values=c("HQ MAG"="#1b9e77", "MQ MAG"="#d95f02"), guide = guide_legend(reverse = TRUE)) 
plot_cov_lr
#saveRDS(plot_cov_lr,"../data/MFD-LR/core_genera.rds")
```


### Calculate coverage ranks per habitat
```{r, include=TRUE}
core_agr_hab <- function(df,type) {
  df$coverage2 <- ifelse(df$coverage == "No MAG" | df$coverage == "SR MAG",0,1)
  df_sum1 <- aggregate(df$coverage2, by=list(df[,6]), FUN=sum)
  colnames(df_sum1) <- c("complex","genera_mag")
  df_sum2 <- aggregate(df$Genus, by=list(df[,6]), FUN=length)
  colnames(df_sum2) <- c("complex","genera_total")
  df_sum <- merge(df_sum1,df_sum2,by="complex")
  df_sum$genera_perc <- round(df_sum$genera_mag/df_sum$genera_total*100,2)
  df_sum$complex0 <- df_sum$complex
  df_sum <- separate(data=df_sum, col=complex0, sep=",", extra="merge",into=c("type", "area", "hab"))
  df_sum$group <- type
  return (df_sum) }

core_area_sum <- core_agr_hab(core_area,"Sample area")
core_hab1_sum <- core_agr_hab(core_hab1,"Habitat (level 1)")
core_hab2_sum <- core_agr_hab(core_hab2,"Habitat (level 2)")
core_hab3_sum <- core_agr_hab(core_hab3,"Habitat (level 3)")

core_hab_sum <- rbind(core_area_sum,core_hab1_sum,core_hab2_sum,core_hab3_sum)
core_hab_sum$type <- factor(core_hab_sum$type,levels =c("Soil","Sediment","Water"))
core_hab_sum$area <- factor(core_hab_sum$area,levels =c(" Agriculture"," Natural"," Urban", " Subterranean"))
```


### Plot core genera coverage at different metadata levels
```{r, include=TRUE}
core_hab_sum <- core_hab_sum[order(core_hab_sum$type,decreasing=TRUE),]

preset <- list(preset_gen2, geom_quasirandom(size = 0.6, width = 0.2),
               geom_violin(scale="width", width=0.25, lwd=0.25,linetype="dashed", fill=NA, alpha=0.5),
               scale_y_continuous(expand=c(0,0), limits = c(0,102),breaks=c(0,20,40,60,80,100)),
               scale_color_manual(values=c("Soil"="#8c510a", "Sediment"="#4d9221", "Water"="#3288bd")))

plot_core_area <- ggplot(data=core_hab_sum[core_hab_sum$group=="Sample area",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for sample area category (%)", x="") + preset

plot_core_hab1 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 1)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 1 (%)", x="") + preset

plot_core_hab2 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 2)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 2 (%)", x="") + preset

plot_core_hab3 <- ggplot(data=core_hab_sum[core_hab_sum$group=="Habitat (level 3)",], aes(x=factor(type, levels=unique(type)), y=genera_perc, color=type)) + 
                  labs(title="", y="Coverage of core genera for habitat level 3 (%)", x="") + preset 

plots_core_hab <- ggarrange(plot_core_area, plot_core_hab1, plot_core_hab2, plot_core_hab3, nrow=2, ncol=2, align = c("v"),
                            legend = "none", common.legend = FALSE, labels = c("a","b","c","d"), font.label = list(size = 9))

#ggsave(plots_core_hab, file="../analysis/figures/FigS9.pdf", height = 90, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Load sylph-related data
```{r, include=TRUE}
# Wrangle metadata
meta <- read.delim("../data/MFD-SR/MFD-SR_metadata_2025-02.tsv", sep="\t", header=T)
meta <- meta[! meta$mfd_sampletype == "Other",]
meta$lib_id <- gsub(".fastq.gz","",meta$flat_name)
meta$mfd_hab1 <- ifelse(meta$mfd_hab1 == "", "Other", meta$mfd_hab1)
meta$mfd_areatype <- ifelse(meta$mfd_areatype == "Agriculture (reclaimed lowland)", "Agriculture", meta$mfd_areatype)
meta_filt <- meta[meta$after_total_bases > 10**9,]

# Function for processing sylph results
sylph_df <- function(df,group,meta) {
    df <- read.delim(df, sep="\t", header=F)
    colnames(df) <- c("lib_id","class")
    df <- merge(df,meta,by="lib_id",all=T)
    df$group <- group
    df <- df[!is.na(df$fieldsample_barcode),]
    df$class[is.na(df$class)] <- 0
    return (df) }

sylph_gtdb <- sylph_df("../data/MFD-SR/sylph-sum-sr-gtdb.tsv","GTDB",meta_filt)
sylph_gtdb_pub <- sylph_df("../data/MFD-SR/sylph-sum-sr-pub.tsv","GTDB + Public catalogues",meta_filt)
sylph_gtdb_pub_lr <- sylph_df("../data/MFD-SR/sylph-sum-sr-pub-mfdlr.tsv","GTDB + Public catalogues + MFD-LR",meta_filt)

sylph_sr <- rbind(sylph_gtdb,sylph_gtdb_pub,sylph_gtdb_pub_lr)
sylph_sr <- sylph_sr %>% group_by(lib_id) %>% filter(n() == 3) %>% ungroup()
median_w_iqr(sylph_sr$class,sylph_sr$group)
```


### Plot classification rates by sample type
```{r, include=TRUE}
sylph_sr$group <- factor(sylph_sr$group,levels = c("GTDB","GTDB + Public catalogues","GTDB + Public catalogues + MFD-LR"))

plot_sylph_sr <- ggplot(data=sylph_sr, aes(x=mfd_sampletype, y=class, by=group, color=group)) + 
                 geom_quasirandom(size = 0.01, alpha=0.4, dodge.width = 0.9) + 
                 geom_boxplot(outlier.shape=NA, fill=NA, width=0.3, alpha = 0.4, lwd=0.1, position = position_dodge(width = 0.9), color="black") + 
                 labs(title="", y="Short reads classified (%)", x="Sample type", color="Database:") + 
                 scale_x_discrete(guide = guide_axis(angle = 0)) + scale_y_continuous(expand=c(0,0), breaks = c(0,25,50,75,100), limits=c(0,102)) + 
                 scale_color_manual(values = c("#e7298a","#66a61e","#e6ab02")) +  preset_gen + 
                 guides(colour = guide_legend(override.aes = list(size=1.5), nrow=3, byrow=TRUE)) +
                 theme(legend.justification = "center", legend.spacing.x = unit(0.1, 'cm'), legend.spacing.y = unit(0.1, 'cm'), 
                       legend.key.height=unit(0.1, 'cm'), legend.key.width=unit(0.1, 'cm'), legend.box.spacing = unit(0, "cm"))

#saveRDS(plot_sylph_sr,"../data/catalogs/sylph.rds")
```


### Plot classification rates by sample type and habitat
```{r, include=TRUE}
preset_sylph <- list(labs(title="", y="Short reads classified (%)", x="", color="Database:"),
                     geom_quasirandom(size = 0.01, alpha=0.4, dodge.width = 0.9),
                     geom_boxplot(outlier.shape=NA, fill=NA, width=0.3, alpha = 0.4, lwd=0.1, position = position_dodge(width = 0.9), color="black"), 
                     scale_x_discrete(guide = guide_axis(angle = 45)),
                     scale_y_continuous(expand=c(0,0), breaks = c(0,25,50,75,100), limits=c(0,102)),
                     scale_color_manual(values = c("#e7298a","#66a61e","#e6ab02")),
                     guides(colour = guide_legend(override.aes = list(size=1.5), nrow=1, byrow=TRUE)),
                     facet_grid(cols = vars(mfd_areatype), scales = "free", space = "free"),
                     preset_gen, theme(strip.text = element_text(size=6), legend.spacing.x = unit(0.1, 'cm'), legend.key.width = unit(0.2, "cm")))

plot_sylph_sr_soil <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Soil",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                      labs(x="Sample habitat (soil sample type)") + preset_sylph + theme(legend.position = "none")

plot_sylph_sr_sed <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Sediment",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                     labs(x="Sample habitat (sediment sample type)") + preset_sylph

plot_sylph_sr_water <- ggplot(data=sylph_sr[sylph_sr$mfd_sampletype == "Water",], aes(x=mfd_hab1, y=class, by=group, color=group)) + 
                       labs(x="Sample habitat (water sample type)") + preset_sylph

fig_sylph_hab_ <- ggarrange(plot_sylph_sr_sed,plot_sylph_sr_water, nrow=1, ncol=2, common.legend = TRUE, legend="bottom", 
                            align = c("h"), labels = c("b","c"), font.label = list(size = 9))

fig_sylph_hab <- ggarrange(plot_sylph_sr_soil,fig_sylph_hab_, nrow=2, ncol=1, common.legend = FALSE, align = c("v"), 
                           labels = c("a",""), font.label = list(size = 9))

#ggsave(fig_sylph_hab, file="../analysis/figures/FigS10.pdf", height = 180, width = 200, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Select sample subset for comparing coastal and agricultural habitats
```{r, include=TRUE}
#meta_sub <- meta[meta$mfd_hab2 != "",]
#meta_sub <- meta_sub[meta_sub$after_total_bases > 10**9 ,]
#meta_sub <- meta_sub[meta_sub$mfd_areatype == "Agriculture" | meta_sub$mfd_hab1 == "Coastal",]

#meta_agr <- meta_sub %>% filter(mfd_areatype == "Agriculture") %>% group_by(across(all_of("mfd_hab2"))) %>% filter(n() >= 50) %>% slice_sample(n = 150) %>% ungroup()
#meta_sub <- rbind(meta_agr,meta_sub[meta_sub$mfd_hab1=="Coastal",])
#write.table(meta_sub,"../data/MFD-SR/meta_sub.tsv",quote=F,row.names=FALSE,col.names=TRUE,sep="\t")
```


### Subset OTU table to selected samples
```{r, include=TRUE}
#otus_all <- read.delim("../data/MFD-SR/2025-02-13_MFD_arcbac_genus_rarefaction_rel.csv", sep=",", header=T)
#subset_mfd <- colnames(otus_all)[colnames(otus_all) %in% meta_sub$fieldsample_barcode]
#otus <- otus_all[, c(subset_mfd, "Kingdom", "Phylum", "Class", "Order", "Family", "Genus")]
#otus <- otus[rowSums(otus[, !(colnames(otus) %in% c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus"))] != 0) > 0, ]
#write.table(otus,"../data/MFD-SR/otu_sub.csv",quote=F,row.names=FALSE,col.names=TRUE,sep=",")
```


### Make distance matrix
```{r, include=TRUE}
#otu_abundance <- otus[, !(colnames(otus) %in% c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus"))]
#otu_abundance <- as.data.frame(lapply(otu_abundance, as.numeric))
#otu_abundance <- t(otu_abundance)

#dist_matrix <- vegdist(otu_abundance, method = "bray")
#saveRDS(dist_matrix, file = "../data/MFD-SR/bray_curtis_dist_matrix.rds")
```


### Reload processed data to improve compute speed
```{r, include=TRUE}
meta_sub <- read.delim("../data/MFD-SR/meta_sub.tsv", sep="\t", header=T)
otus <- read.delim("../data/MFD-SR/otu_sub.csv", sep=",", header=T)
dist_matrix <- readRDS("../data/MFD-SR/bray_curtis_dist_matrix.rds")
```


### Get pcoa
```{r, include=TRUE}
pcoa_result <- pcoa(dist_matrix)
pcoa_df <- as.data.frame(pcoa_result$vectors[, 1:3])
colnames(pcoa_df) <- c("pcoa_1","pcoa_2","pcoa_3")
pcoa_df$fieldsample_barcode <- rownames(pcoa_df)

meta_sub_ <- merge(meta_sub,pcoa_df,by="fieldsample_barcode")
```


### Run ANOSIM tests:  coastal/agriculture (R=0.7554, p=0.001), coastal (R=0.3288, p=0.001)
```{r, include=TRUE}
# Variance explained
pcoa_1 <- round(sum(as.vector(pcoa_result$value$Relative_eig)[1])*100,1)
pcoa_2 <- round(sum(as.vector(pcoa_result$value$Relative_eig)[2])*100,1)
pcoa_3 <- round(sum(as.vector(pcoa_result$value$Relative_eig)[3])*100,1)

# ANOSIM - coastal vs agriculture
dist_samples <- labels(dist_matrix)
meta_sub_ <- meta_sub_[match(dist_samples, meta_sub_$fieldsample_barcode), ]
meta_samples <- meta_sub_$fieldsample_barcode
all(dist_samples == meta_samples) # If TRUE, then the categories match correctly

#grouping_vector <- as.factor(meta_sub_$mfd_hab1)
#anosim_result <- anosim(dist_matrix, grouping_vector)
#summary(anosim_result)

# ANOSIM - coastal habitats
meta_sub_2 <- meta_sub_[meta_sub_$mfd_hab1=="Coastal",]
sub2 <- dist_samples[dist_samples %in% meta_sub_2$fieldsample_barcode]
dist_matrix2 <- as.matrix(dist_matrix)[sub2, sub2]
all(sub2 == meta_sub_2$fieldsample_barcode)

#grouping_vector <- as.factor(meta_sub_2$mfd_hab2)
#anosim_result2 <- anosim(dist_matrix2, grouping_vector)
#summary(anosim_result2)
```


### Plot PCoA
```{r, include=TRUE}
count_samples <- function(df,group) {
  samples_n <- aggregate(df[,group], by=list(df[,group]), FUN=length)
  colnames(samples_n) <- c("group","n")
  samples_n$n_ <- paste0(samples_n$group," (",samples_n$n,")")
  samples_n <- samples_n %>% select(group, n_) %>% distinct() %>% deframe()
  return (samples_n) }

labels_n1 <- count_samples(meta_sub_,"mfd_hab1")
labels_n2 <- count_samples(meta_sub_,"mfd_hab2")

colors_hab2 <- c("Salt marshes and salt meadows"="#1d91c0",
                 "Sea cliffs and shingle or stony beaches"="#0c2c84",
                 "Poales, grass"="#8c2d04", "Poales, Cereal"="#ec7014",
                 "Asterids"="#fec44f", "Malvids"="#ffd92f",
                 "Fallow"="#dfc27d", "Mixed crops"="#8c510a")

plot_pcoa <- ggplot(meta_sub_, aes(x = pcoa_1, y = pcoa_2, color = mfd_hab2)) +
             stat_ellipse(aes(fill = mfd_hab1), alpha = 0.1, geom = "polygon") + geom_point(size = 0.7, alpha=0.9) +
             scale_fill_manual(labels=labels_n1, values=c("Coastal"="#a6cee3","Fields"="#b2df8a")) +
             scale_color_manual(labels=labels_n2, values=colors_hab2) +
             labs(x = paste0("PCoA1 - ",pcoa_1," %"), 
                  y = paste0("PCoA2 - ",pcoa_2," %"), 
                  color="Habitat descriptor (level 2):", 
                  fill="Habitat descriptor (level 1):") + preset_gen +
             guides(color = guide_legend(title.position = "top", override.aes = list(fill = NA, size=1.5)), 
                    fill = guide_legend(title.position = "top", ncol=1, order=1)) +
             theme(legend.key.size = unit(6, 'mm'), legend.spacing.x = unit(0.1, 'cm'), legend.spacing.y = unit(0.1, 'cm'),
                   legend.key.height=unit(0.5, 'cm'), legend.key.width=unit(0.5, 'cm'), legend.box.spacing = unit(0, "cm"),
                   legend.box.background = element_blank(),legend.background = element_blank())
plot_pcoa
```


### Plot sample map
```{r, include=TRUE}
map <- mapDK(detail = 'region', map.colour = "grey50", map.fill = "grey98") + 
       geom_point(data = meta_sub, aes(x=longitude, y=latitude, group=mfd_hab2, color=mfd_hab2), size=0.5)  +
       scale_color_manual(labels=labels_n2, values=colors_hab2) +
       guides(colour = guide_legend(override.aes = list(size=1.2), ncol=1, title.position = "top", reverse=TRUE, title="Habitat descriptor (level 2):")) 
       theme(legend.key=element_blank(), legend.position = c(0.75, 0.85),
             legend.box.background = element_blank(), legend.background = element_blank(),
             legend.text=element_text(size=7),legend.title = element_text(size=7),
             legend.key.size = unit(3.5, 'mm'), legend.spacing.x = unit(1, 'mm'),
             plot.margin = margin(-5, -5, -5, -5))
map
```


### Get phylum-level abundances
```{r, include=TRUE}
otus_p <- otus %>% group_by(Phylum) %>% summarise(across(where(is.numeric), sum, na.rm = TRUE))

otus_p <- otus_p %>% pivot_longer(cols = -Phylum, names_to = "fieldsample_barcode", values_to = "abund")
otus_p <- otus_p[otus_p$Phylum != "Unclassified",]
otus_p <- merge(otus_p,meta_sub[,c("fieldsample_barcode","mfd_hab1","mfd_hab2")],by="fieldsample_barcode")

otus_p_all <- aggregate(otus_p$abund, by=list(otus_p$Phylum), FUN=median)
otus_p_all <- otus_p_all[order(otus_p_all$x,decreasing=TRUE),]

otus_p_med <- aggregate(otus_p$abund, by=list(otus_p$Phylum,otus_p$mfd_hab2), FUN=median)
colnames(otus_p_med) <- c("Phylum","mfd_hab2","abund_med")
otus_p_med <- otus_p_med[otus_p_med$Phylum %in% head(otus_p_all$Group.1,10),]
otus_p_med <- merge(otus_p_med,meta_sub[,c("mfd_hab1","mfd_hab2")],by="mfd_hab2")
otus_p_med$abund_med <- round(otus_p_med$abund_med,1)
```


### Phylum median abundance plot
```{r, include=TRUE}
otus_p_med <- otus_p_med %>% mutate(Phylum = factor(Phylum, levels = rev(c(otus_p_all$Group.1)))) %>% arrange(Phylum)

plot_abund_phyl <- ggplot(data = otus_p_med, mapping = aes(x = mfd_hab2, y = Phylum, fill = abund_med)) + geom_tile(width=1) +
                   geom_text(aes(label = abund_med), size=1.75, fontface = "plain") +
                   labs(title="", y="Phylum", x="", fill="Median relative abundance (%):") + 
                   scale_x_discrete(guide = guide_axis(angle = 30), expand = c(0, 0)) + scale_y_discrete(expand = c(0, 0)) +
                   scale_fill_gradient2(midpoint = 9.5, limits=c(0,20), breaks=c(0,5,10,15,20),
                                        low = "#b2182b", mid= "#fee08b", high = "#3288bd", na.value = "#3288bd") + preset_gen +
                   facet_grid(cols = vars(factor(mfd_hab1, levels = (unique(mfd_hab1)))), scales = "free", space = "free") +
                   guides(fill = guide_colorbar(title.position = "top")) +
                   theme(strip.text = element_text(size=6), plot.margin = margin(-15, 5, -0.5, 3),
                         legend.box.background = element_blank(), legend.background = element_blank(),
                         legend.key.height = unit(3.5, "mm"), legend.key.width = unit(6.5, "mm"),
                         legend.direction = "horizontal")
plot_abund_phyl
```


### Abundance plots for select phyla
```{r, include=TRUE}
preset_hab <- list(geom_quasirandom(size = 0.01, width = 0.3), geom_violin(lwd=0.25,linetype="dashed", fill=NA), preset_gen,
                   scale_x_discrete(guide = guide_axis(angle = 30), expand = c(0, 0)),
                   facet_grid(cols = vars(factor(mfd_hab1, levels = rev(unique(mfd_hab1)))), scales = "free", space = "free") ,
                   theme(strip.text = element_text(size=6), plot.margin = margin(-15, 2, -0.5, 2)),
                   scale_y_continuous(trans="log2"),
                   scale_color_manual(values=colors_hab2))

plot_phyl1 <- ggplot(data=otus_p[otus_p$Phylum=="Firmicutes",], aes(x=mfd_hab2, y=abund, color=mfd_hab2)) + 
              labs(title="", y="Firmicutes\nabundance (%)", x="") +  preset_hab + theme(axis.text.x = element_blank())
plot_phyl1

plot_phyl2 <- ggplot(data=otus_p[otus_p$Phylum=="Proteobacteria",], aes(x=mfd_hab2, y=abund, color=mfd_hab2)) + 
              labs(title="", y="Proteobacteria\nabundance (%)", x="") + preset_hab + theme(axis.text.x = element_blank()) + coord_cartesian(ylim = c(2, 60))
plot_phyl2

plot_phyl3 <- ggplot(data=otus_p[otus_p$Phylum=="Bacteroidota",], aes(x=mfd_hab2, y=abund, color=mfd_hab2)) + 
              labs(title="", y="Bacteroidota\nabundance (%)", x="") +  preset_hab
plot_phyl3
```


### Assemble plots
```{r, include=TRUE}
# First half
plot_habs_ab <- ggarrange(map,plot_pcoa,widths=c(1,1), nrow=1, ncol=2, align = "none", common.legend=FALSE, legend = "none",
                          labels = c("a","b"), font.label = list(size = 9))

plot_habs_ab_leg <- as_ggplot(get_legend(plot_pcoa)) + theme(plot.margin = margin(-60, 0, 0, 0))
plot_habs_ab <- ggarrange(plot_habs_ab, plot_habs_ab_leg, nrow=2, align = "none", heights=c(1,0.4)) + theme(plot.margin = margin(0, 0, -60, 0))

# Second half
plot_df <- ggarrange(plot_phyl1,plot_phyl2,plot_phyl3, nrow=3, ncol=1, heights=c(1,1,1.75), align = "v", common.legend=FALSE, legend = "none",
                     labels = c("d","e","f"), font.label = list(size = 9))

plot_cf <- ggarrange(plot_abund_phyl,plot_df, nrow=1, ncol=2, align = "hv", common.legend=FALSE, legend="none",
                     labels = c("c",""), font.label = list(size = 9), widths=c(1,1.1)) + theme(plot.margin = margin(5, 0, 0, 0))

plot_cf_leg <- as_ggplot(get_legend(plot_abund_phyl)) + theme(plot.margin = margin(-110, 0, 0, 0))
plot_cf <- ggarrange(plot_cf, plot_cf_leg, nrow=2, align = "v", heights=c(1,0.2)) + theme(plot.margin = margin(0, 0, -55, 0))

plot_af <- ggarrange(plot_habs_ab,plot_cf, nrow=2, ncol=1, align = "v", common.legend=FALSE, legend = "none") + theme(plot.margin = margin(0, 0, -10, 0))

#ggsave(plot_af, file="../analysis/figures/FigS5.pdf", height = 200, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```

