---
title: "R code for plotting MFD-LR MAG catalog stats"
author: "Mantas Sereika"
date: "2024"
output:
  html_document: default
  pdf_document: default
---


###  Load dependencies
```{r, include=TRUE, message=FALSE}
library(stringr)
library(stringi)
library(tidyverse)
library(dplyr)
library(ggplot2)
library(ggpubr)
library(mapDK)
library(iNEXT)
library(ggsankey)
library(ggbeeswarm)
library(grid)
library(ggplotify)
library(lemon)
library(pals)
library(gridExtra)
```


### Load data
```{r, include=TRUE}
gen <- read.delim("../analysis/datasets/dataset_S1.tsv", sep="\t", header=T)
mags <- read.delim("../analysis/datasets/dataset_S3.tsv", sep="\t", header=T)
```


### Wrangle metadata
```{r, include=TRUE}
# Rename outlier groups
gen$mfd_hab1_ <-  gen$mfd_hab1
gen$mfd_hab1_ <- ifelse( gen$fieldsample_barcode == "MFD06085", "Other",  gen$mfd_hab1_)
gen$mfd_hab1_ <- ifelse( gen$fieldsample_barcode == "MFD03250", "Other",  gen$mfd_hab1_)
gen$mfd_hab1_ <- ifelse( gen$fieldsample_barcode == "MFD09603", "Other",  gen$mfd_hab1_)
gen$mfd_hab1_ <- ifelse( gen$fieldsample_barcode == "MFD10097", "Other",  gen$mfd_hab1_)

# Fill in missing values
gen$mfd_hab2 <- ifelse(gen$mfd_hab2 == "", "Unassigned habitat (level 2)", gen$mfd_hab2)
gen$mfd_hab3 <- ifelse(gen$mfd_hab3 == "", "Unassigned habitat (level 3)", gen$mfd_hab3)

# Get counts per habitat
hab_count <- aggregate(gen$mfd_hab1_, by=list(Category=gen$mfd_hab1_), drop=FALSE, FUN=length)
colnames(hab_count) <- c("mfd_hab1_","n")
hab_count$mfd_hab1_n <- paste0(hab_count$mfd_hab1_," (",hab_count$n,")")
gen <- merge(gen,hab_count[,c("mfd_hab1_","mfd_hab1_n")],by="mfd_hab1_")

# Update MAG dataframe
mags[,c("mfd_sampletype","mfd_areatype","mfd_hab1","mfd_hab2","mfd_hab3")] <- NULL
mags <- merge(mags,gen[,c("fieldsample_barcode","mfd_sampletype","mfd_areatype","mfd_hab1","mfd_hab1_","mfd_hab2","mfd_hab3")],by="fieldsample_barcode")
```


### Plot presets
```{r, include=TRUE}
colors <- c("#85660D","#FEAF16","#690726","#90AD1C","#1C8356","#4eb3d3",
            "#1CBE4F","#782AB6","#DEA0FD","#325A9B","#C4451C","#F8A19F" )

labels <- c("Bogs, mires and fens"="#85660D",
            "Coastal"="#FEAF16",
            "Dunes"="#690726",
            "Fields"="#90AD1C",
            "Forests"="#1C8356",
            "Urban freshwater"="#4eb3d3",
            "Natural freshwater"="#0080FE",
            "Grassland formations"="#1CBE4F",
            "Natural, unassigned habitat"="#782AB6",
            "Rocky habitats and caves"="#DEA0FD",
            "Saltwater"="#325A9B",
            "Temperate heath and scrub"="#C4451C",
            "Greenspaces"="#F8A19F",
            "Soil"="#7c4807",
            "Sediment"="#b3b3b3",
            "Water"="#0080FE",
            "Other"="#DEA0FD")

labels_n <- gen %>% select(mfd_hab1_, mfd_hab1_n) %>% distinct() %>% deframe()

preset_gen1 <- list(theme_bw(),
                    theme(legend.position = "none",
                           axis.title.y = element_text(size = 7), 
                           axis.title.x = element_text(size = 7),
                           axis.text.y = element_text(size = 6),
                           axis.text.x = element_text(size = 6),
                           panel.grid.minor = element_line(size = 0.15),
                           panel.grid.major = element_line(size = 0.3)),
                    scale_x_discrete(guide = guide_axis(angle = 45)),
                    scale_color_manual(values=labels))

preset_gen2 <- list(theme_bw(),
                    theme(legend.position = "bottom",
                          legend.text=element_text(size=7),
                          legend.title = element_text(size=7),
                           axis.title.y = element_text(size = 7), 
                           axis.title.x = element_text(size = 7),
                           axis.text.y = element_text(size = 6),
                           axis.text.x = element_text(size = 6),
                           panel.grid.minor = element_line(size = 0.15),
                           panel.grid.major = element_line(size = 0.3)))

preset_gen3 <- list(theme_bw(),
                    theme(legend.position = "none",
                           axis.title.y = element_text(size = 7), 
                           axis.title.x = element_text(size = 7),
                           axis.text.y = element_text(size = 6),
                           axis.text.x = element_text(size = 6),
                           axis.ticks = element_blank(),
                           panel.grid.minor = element_line(size = 0.15),
                           panel.grid.major = element_line(size = 0.3)),
                     coord_flip())
```


### Function for getting medians and IQR
```{r, include=TRUE}
median_w_iqr <- function(x,cat) {
  df_med <- aggregate(x, by=list(cat), FUN=median, na.rm = T)
  colnames(df_med) <- c("group","median")
  
  df_iqr <- aggregate(x, by=list(cat), FUN=IQR, na.rm = T)
  colnames(df_iqr) <- c("group","iqr")
  
  df_q25 <- aggregate(x, by=list(cat), FUN = function(i,...) quantile(i, probs=0.25, na.rm=T))
  colnames(df_q25) <- c("group","q25")
  
  df_q75 <- aggregate(x, by=list(cat), FUN = function(i,...) quantile(i, probs=0.75, na.rm=T))
  colnames(df_q75) <- c("group","q75")
  
  df <- merge(df_med,df_iqr,by="group")
  df <- merge(df,df_q25,by="group")
  df <- merge(df,df_q75,by="group")
  return (df) }
```


### Function for splitting GTDB ranks
```{r, include=TRUE}
gtdb_rank <- function(class,pattern,trim_left,trim_right) {
  rank <-  str_extract(class,pattern)
  rank <-  gsub(trim_left, "", rank)
  rank <-  gsub(trim_right, "", rank)
  return (rank) }
```


### Sankey for sample MFD metadata
```{r, include=TRUE}
# Set colors
samples_soil <- gen[gen$mfd_sampletype == "Soil",]
samples_soil_n <- c(unique(samples_soil$mfd_sampletype), unique(samples_soil$mfd_hab1), unique(samples_soil$mfd_hab2), unique(samples_soil$mfd_hab3))
samples_soil_col <- rev(brewer.greens(length(samples_soil_n)))

samples_other <- gen[! gen$mfd_sampletype == "Soil",]
samples_other_n <- c(unique(samples_other$mfd_sampletype), unique(samples_other$mfd_hab1),unique(samples_other$mfd_hab2), unique(samples_other$mfd_hab3))
samples_other_col <- rev(brewer.blues(length(samples_other_n)))

samples <- unique(c(unique(gen$mfd_areatype),samples_soil_n,samples_other_n))
samples_col <- head(c(c("#f0d851","#ff9900","#fb4f4f","#360000"),samples_soil_col,samples_other_col),length(samples))

# Initial dataframe for ggsankey
gen_sankey <- gen %>% select(mfd_sampletype,mfd_areatype,mfd_hab1,mfd_hab2,mfd_hab3) %>% make_long(mfd_areatype,mfd_sampletype,mfd_hab1,mfd_hab2,mfd_hab3)

# Add group counts
gen_sankey_sum <- gen_sankey %>% dplyr::group_by(node) %>% tally()
gen_sankey_ <- merge(gen_sankey,gen_sankey_sum, by.x = 'node', by.y = 'node', all.x = TRUE)

# Filter
gen_sankey_  <- gen_sankey_ %>% filter(!is.na(node))

# Order
gen_sankey_sum <- gen_sankey_sum[order(gen_sankey_sum$n,decreasing=FALSE),]
gen_sankey_$node <- factor(gen_sankey_$node,levels=rev(samples))
gen_sankey_$next_node <- factor(gen_sankey_$next_node,levels=rev(samples))

sankey <- ggplot(gen_sankey_, aes(x = x, next_x = next_x, node = node, next_node = next_node,
                                  fill = factor(node), label = paste0(node," (", n,")"))) +  
          geom_sankey(flow.alpha = 0.5, node.color = "black", size=0.05, show.legend = FALSE, width = 0.1) +
          geom_sankey_label(size = 1.2, color = "black", fill= "white",hjust=0, alpha=0.85, width=2) +  theme_bw() +
          theme(axis.title = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank(),
                panel.grid = element_blank(), legend.position = "none",
                axis.text.x = element_text(margin = margin(t = -10),size = 7), panel.border = element_blank(),
                plot.margin=grid::unit(c(-2,2,1,-15), "mm")) +
          scale_fill_manual(values=samples_col,breaks=samples) + 
          scale_x_discrete(labels=c("mfd_sampletype" = "Sample type", 
                                    "mfd_areatype" = "Area type", 
                                    "mfd_hab1" = "Habitat descriptor\n(level 1)",
                                    "mfd_hab2" = "Habitat descriptor\n(level 2)",
                                    "mfd_hab3" = "Habitat descriptor\n(level 3)"))

sankey
ggsave(file="../analysis/figures/FigS1.pdf", height = 180, width = 150, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Map of sequenced samples
```{r, include=TRUE}
map <- mapDK(detail = 'region', map.colour = "grey50", map.fill = "grey98") + 
       geom_point(data = gen, aes(x = longitude, y = latitude, group = NA, color = mfd_hab1_), size = 1) +
       theme(legend.key=element_blank(), legend.position = c(0.85, 0.75),
             legend.box.background = element_blank(), legend.background = element_blank(),
             legend.text=element_text(size=7),legend.title = element_text(size=7),
             legend.key.size = unit(3.5, 'mm'), legend.spacing.x = unit(1, 'mm')) +
       labs(col="Habitat:") +  scale_color_manual(values=labels, labels = labels_n) +
       guides(colour = guide_legend(override.aes = list(size=1.2), ncol=1, title.position = "top")) 
```


### General per-sample metrics
```{r, include=TRUE}
gen$name1 <- ""

preset_fig1 <- list(geom_quasirandom(size = 0.2, width = 0.2, aes(color=mfd_hab1_)),
                    scale_color_manual(values=labels),
                    geom_boxplot(outlier.shape=NA, width=0.6, alpha = 0.4, lwd=0.2), 
                    preset_gen3, theme(plot.margin = unit(c(-0.75,0.3,0.05,-0.2), 'lines')) )

plot_scale <- scale_y_continuous(expand=c(0,0), limits = c(0,102), breaks = c(0,25,50,75,100))

plot_yield <- ggplot(data=gen, aes(x=name1, y=reads_size_bp/10**9)) + labs(title="", y="Sequenced yield (Gbp)", x="") +
              preset_fig1 + scale_y_continuous(expand=c(0,0), limits = c(0,210), breaks = c(0,50,100,150,200))

plot_n50 <- ggplot(data=gen, aes(x=name1, y=reads_N50_bp/10**3)) + labs(title="", y="Read N50 (kbp)", x="") +
            preset_fig1 + scale_y_continuous(expand=c(0,0), limits = c(0,12.5), breaks = c(0,3,6,9,12))

plot_asm <- ggplot(data=gen, aes(x=name1, y=assembled)) + labs(title="", y="Sequenced data asssembled (%)", x="") +
            preset_fig1 + plot_scale

plot_bin <- ggplot(data=gen, aes(x=name1, y=r_abund*(assembled/100))) + labs(title="", y="Sequenced data binned (%)", x="") +
            preset_fig1 + plot_scale

plot_mags <- ggplot(data=gen, aes(x=name1, y=HQ_mags+MQ_mags)) + labs(title="", y="MAGs recovered", x="")  +
             preset_fig1 + scale_y_continuous(expand=c(0,0), limits = c(0,420), breaks = c(0,100,200,300,400))
```


### Initial Figure 1
```{r, include=TRUE}
plot_leg <- get_legend(map)
plot_leg <- as_ggplot(plot_leg) + theme(plot.margin = unit(c(0, 0.5, 1.25, 0.5), "cm"))

fig1_a <- ggarrange(map, plot_leg, nrow=2, ncol=1, legend = "none", common.legend = FALSE, labels = "a", font.label = list(size = 9), heights=c(3.5,1))

fig1_bf <- ggarrange(plot_yield, plot_n50, plot_asm, plot_bin, plot_mags,
                     nrow=5, ncol=1, legend = "none", common.legend = FALSE,
                     labels = c("b","c","d","e","f"), font.label = list(size = 9))

fig1 <- ggarrange(map, fig1_bf, nrow=1, ncol=2, common.legend = FALSE, widths=c(2,1), align = c("v"), labels = c("a"), font.label = list(size = 9))
fig1

ggsave(file="../analysis/figures/Fig1_pre.pdf", height = 100, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### General per-sample metrics by habitat
```{r, include=TRUE}
preset_hab <- list(geom_quasirandom(size = 0.65, width = 0.2), geom_violin(lwd=0.25,linetype="dashed", fill=NA), preset_gen1)

plot_gb <- ggplot(data=gen, aes(x=mfd_hab1_, y=reads_size_bp/10**9, color=mfd_hab1_)) + 
           labs(title="", y="Sequenced yield (Gbp)", x="") + preset_hab

plot_asm <- ggplot(data=gen, aes(x=mfd_hab1_, y=assembled, color=mfd_hab1_)) + 
            labs(title="", y="Data assembled (%)", x="") + preset_hab

plot_count <- ggplot(data=gen, aes(x=mfd_hab1_, y=(HQ_mags+MQ_mags), color=mfd_hab1_)) + 
              labs(title="", y="MAGs recovered (HQ and MQ)", x="") + preset_hab +
              scale_y_continuous(expand=c(0,0), limits = c(0,470),breaks=c(0,100,200,300,400)) 

plot_bind <- ggplot(data=gen, aes(x=mfd_hab1_, y=round(r_abund*assembled/100,0), color=mfd_hab1_)) + 
             labs(title="", y="Data binned in MAGs (%)", x="") + preset_hab +
             scale_y_continuous(expand=c(0,0), limits = c(0,70),breaks=c(0,20,40,60)) 

plots_mag_hab <- ggarrange(plot_gb, plot_asm, plot_count, plot_bind, nrow=2, ncol=2, align = c("v"),
                           legend = "none", common.legend = FALSE, labels = c("a","b", "c", "d"), font.label = list(size = 9))
plots_mag_hab  

ggsave(file="../analysis/figures/FigS3.pdf", height = 150, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Add homopolymer counts to MAG dataframe
```{r, include=TRUE}
hp <- read.delim("../data/MFD-LR/hp.txt", sep=" ", header=F)
colnames(hp) <- c("bin","homopol","count")
hp$bin <- gsub(".fa", "", hp$bin)

hp$nuc <- str_extract(hp$homopol, "^.{1}")
hp$homopol = substring(hp$homopol, 2)
hp$homopol = as.integer(hp$homopol)

hp_long <- hp[(hp$homopol >= 7), ]
hp_long_sum <- aggregate(hp_long$count, by=list(hp_long$bin), FUN=sum, drop=FALSE)
colnames(hp_long_sum) <- c("bin","hp_long")

mags <- merge(mags, hp_long_sum, by="bin", all=TRUE)
mags$hp_long[is.na(mags$hp_long)] <- 0
```


### Wrangle MAG dataframe
```{r, include=TRUE}
mags$domain <- gtdb_rank(mags$GTDB_taxonomy,"d__[a-zA-Z0-9 -_]+;p__","d__",";p__")
mags$phylum <- gtdb_rank(mags$GTDB_taxonomy,";p__[a-zA-Z0-9 -_]+;c__",";p__",";c__")
mags$class <- gtdb_rank(mags$GTDB_taxonomy,";c__[a-zA-Z0-9 -_]+;o__",";c__",";o__")
mags$order <- gtdb_rank(mags$GTDB_taxonomy,";o__[a-zA-Z0-9 -_]+;f__",";o__",";f__")
mags$family <- gtdb_rank(mags$GTDB_taxonomy,";f__[a-zA-Z0-9 -_]+;g__",";f__",";g__")
mags$genus <- gtdb_rank(mags$GTDB_taxonomy,";g__[a-zA-Z0-9 -_]+;s__",";g__",";s__")
mags$species <- gtdb_rank(mags$GTDB_taxonomy,";s__[a-zA-Z0-9 -_]+",";s__","")

mags$phylum[is.na(mags$phylum)] <- paste0("Unclassified")
mags$class[is.na(mags$class)] <- paste0("Unclassified")
mags$order[is.na(mags$order)] <- paste0("Unclassified")
mags$family[is.na(mags$family)] <- paste0("Unclassified")
mags$genus[is.na(mags$genus)] <- paste0("Unclassified")
mags$species[is.na(mags$species)] <- paste0("Unclassified")

mags <- mags[order(mags$domain,decreasing=TRUE),]
mags <- mags[order(mags$MAG_status2,decreasing=TRUE),]
```


### Plot dereplicated MAG metrics
```{r, include=TRUE}
preset_drep <- list(geom_point(alpha=0.5, size=0.8, stroke=0.15, pch=21), preset_gen2,
                    scale_fill_manual(name = "MAG quality:", values = c("cMAG" = "#045a8d","HQ"="#a6dba0","MQ" = "#fee090")),
                    scale_color_manual(name = "Domain:", values=c("Bacteria"="gray90", "Archaea"="#99000d")),
                    guides(colour = guide_legend(override.aes = list(size=2, stroke=0.75, alpha=0.75)), fill = guide_legend(override.aes = list(size=2))))

plot_cm2 <- ggplot(data=mags[mags$rep_status == TRUE,], aes(x=Completeness_CheckM2,y=Contamination_CheckM2,fill=MAG_status2, col=domain)) + 
            scale_x_continuous(breaks=c(50,60,70,80,90,100), expand=c(0,0), limits = c(49,101)) +
            scale_y_continuous(breaks=c(0,2,4,6,8,10), expand=c(0,0), limits = c(-0.25,10.25)) +
            labs(x="Completeness (%)",y="Contamination (%)") + preset_drep

plot_n50_cov  <- ggplot(data=mags[mags$rep_status == TRUE,],aes(x=cov,y=Contig_N50/10**3,col=domain, fill=MAG_status2)) +
                 scale_x_continuous(trans="log2", breaks=c(10,40,160,640), expand=c(0,0), limits = c(2.5,1350)) +
                 scale_y_continuous(trans="log2") + labs(x="MAG coverage",y="Contig N50 (kbp)") + preset_drep

plot_cd_cov <- ggplot(data=mags[mags$rep_status == TRUE,],aes(x=cov,y=Coding_Density*100,col=domain, fill=MAG_status2)) +   
               scale_x_continuous(trans="log2", breaks=c(10,40,160,640), expand=c(0,0), limits = c(2.5,1350)) +
               labs(x="MAG coverage",y="Coding density (%)") + preset_drep

plot_cd_gc <- ggplot(data=mags[mags$rep_status == TRUE,],aes(x=GC_Content,y=Coding_Density*100,col=domain, fill=MAG_status2)) +
              labs(x="GC content (%)",y="Coding density (%)") + preset_drep

plot_cd_hp <- ggplot(data=mags[mags$rep_status == TRUE,],aes(x=hp_long/(Genome_Size/10**5) ,y=Coding_Density*100,col=domain, fill=MAG_status2)) + 
              scale_x_continuous(trans="log2") + labs(x="Long homopolymers (>6) per 100 kbp",y="Coding density (%)") + preset_drep

plot_gc_hp <- ggplot(data=mags[mags$rep_status == TRUE,],aes(x=hp_long/(Genome_Size/10**5) ,y=GC_Content,col=domain, fill=MAG_status2)) +
              scale_x_continuous(trans="log2") + labs(x="Long homopolymers (>6) per 100 kbp",y="GC content (%)") + preset_drep

plot_mag_drep <- ggarrange(plot_cm2,plot_n50_cov,plot_cd_cov,plot_cd_gc,plot_cd_hp,plot_gc_hp, nrow=3, ncol=2, 
                           legend = "bottom", common.legend = TRUE, labels = c("a","b","c","d","e","f"), font.label = list(size = 9))

ggsave(file="../analysis/figures/FigS5.pdf", height = 200, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Set sankey plot for MAG recovery
```{r, include=TRUE}
mags$sankey1 <- ifelse(mags$MAG_status2 == "cMAG", "Circular",
                       ifelse(mags$Total_Contigs == 1, "Single-contig",
                              ifelse(mags$Total_Contigs <= 10,"10 or less contigs",
                                     ifelse(mags$Total_Contigs <= 25,"25 or less contigs",
                                            ifelse(mags$Total_Contigs <= 50,"50 or less contigs","More than 50 contigs")))))

mags$sankey2 <- paste("",mags$MAG_status, sep="")
mags$sankey3 <- ifelse(mags$rep_status == TRUE, paste("Dereplicated\n",mags$MAG_status,sep=""), NA)

mags$gtdb_status <- ifelse(mags$phylum == "Unclassified", "Unassigned phyla", 
                           ifelse(mags$class == "Unclassified", "Unassigned class",
                                  ifelse(mags$order == "Unclassified", "Unassigned order",
                                         ifelse(mags$family == "Unclassified", "Unassigned family",
                                                ifelse(mags$genus == "Unclassified", "Unassigned genus",
                                                       ifelse(mags$species == "Unclassified", "Unassigned species", "Known species"))))))

mags$sankey4 <- ifelse(mags$rep_status == TRUE, mags$gtdb_status, NA)
mags$sankey4 <- ifelse(mags$sankey4 == "Unassigned species", "Unassigned\nspecies", mags$sankey4)
```


### Sankey plot for MAG stats
```{r, include=TRUE}
# Initial df for ggsankey
mags_sankey <- mags %>% select(sankey1,sankey2,sankey3,sankey4) %>% make_long(sankey1,sankey2,sankey3,sankey4)

# Add group counts
mags_sankey_sum <- mags_sankey %>% dplyr::group_by(node) %>% tally()
mags_sankey_ <- merge(mags_sankey,mags_sankey_sum, by.x = 'node', by.y = 'node', all.x = TRUE)

# Filter
mags_sankey_  <- mags_sankey_ %>% filter(!is.na(node))

# Ordering
mags_sankey_$node <- factor(mags_sankey_$node,levels =
                              c("More than 50 contigs","50 or less contigs","25 or less contigs","10 or less contigs",
                                "Single-contig", "Circular",
                                "MQ","HQ","cMAG",
                                "Redundant","Dereplicated\nMQ","Dereplicated\nHQ","Dereplicated\ncMAG",
                                "Omitted", "Known species","Unassigned\nspecies", "Unassigned genus",
                                "Unassigned family","Unassigned order","Unassigned class"))

mags_sankey_$next_node <- factor(mags_sankey_$next_node,levels =
                              c("More than 50 contigs","50 or less contigs","25 or less contigs","10 or less contigs",
                                "Single-contig", "Circular",
                                "MQ","HQ","cMAG",
                                "Redundant","Dereplicated\nMQ","Dereplicated\nHQ","Dereplicated\ncMAG",
                                "Omitted", "Known species","Unassigned\nspecies", "Unassigned genus",
                                "Unassigned family","Unassigned order","Unassigned class"))

sankey <- ggplot(mags_sankey_, aes(x = x, next_x = next_x, node = node, next_node = next_node, fill = factor(node), label = paste0(node," (", n,")"))) +  
          geom_sankey(flow.alpha = 0.5, node.color = "black", size=0.2, show.legend = FALSE, width = 0.1) +
          geom_sankey_label(size = 1.75, color = "black", fill= "white",hjust=0, alpha=0.85, width=2.25) +  theme_bw() +
          theme(axis.title = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank(),
                panel.grid = element_blank(), legend.position = "none", panel.border = element_blank(),
                axis.text.x = element_text(margin = margin(t = -10),size = 7), plot.margin=grid::unit(c(0,0,0,-18), "mm")) +
          scale_fill_viridis_d(option = "inferno", direction=-1) +
          scale_x_discrete(labels=c("sankey1" = "MAG\ncontiguity", "sankey2" = "Quality ranking\n(all MAGs)",
                                    "sankey3" = "Quality ranking\n(dereplicated)", "sankey4" = "Taxonomy\nranking"))

sankey
```


### Plot general MAG stats:
```{r, include=TRUE}
preset_mags <- list(geom_boxplot(width=0.75, alpha = 0.75, lwd=0.2, outlier.size=0.1), scale_x_discrete(limits=rev),
                    preset_gen3,theme(plot.margin = unit(c(-2, 0, 0, -2), "mm")),
                    scale_fill_manual(name = "MAG quality:", values = c("cMAG" = "#045a8d","HQ"="#1b9e77","MQ" = "#d95f02")))

plot_mag_n50 <- ggplot(data=mags[mags$rep_status == TRUE,], aes(x=MAG_status, y=Contig_N50/10**3, fill=MAG_status)) +
                preset_mags + labs(title="", y="Contig N50 (kbp)", x="") + scale_y_continuous(trans="log2") 

plot_mag_cov <- ggplot(data=mags[mags$rep_status == TRUE,], aes(x=MAG_status, y=as.numeric(cov), fill=MAG_status)) + 
                preset_mags + labs(title="", y="MAG coverage", x="") + scale_y_continuous(trans="log2") 

plot_mag_cd <- ggplot(data=mags[mags$rep_status == TRUE,], aes(x=MAG_status, y=Coding_Density*100, fill=MAG_status)) + 
               preset_mags + labs(title="", y="Coding density (%)", x="") 
 
plot_mag_size <- ggplot(data=mags[mags$rep_status == TRUE,], aes(x=MAG_status, y=Genome_Size/10**6, fill=MAG_status)) + 
                 preset_mags + labs(title="", y="MAG size (Mbp)", x="")
```


### Plot species-level rarefaction curve
```{r, include=TRUE}
drep_rf <- aggregate(mags$drep_cluster, by=list(mags$drep_cluster), FUN=length)

drep_rf_ <- iNEXT(drep_rf[,c("x")], q=0, datatype="abundance")

drep_rf_plot <- ggiNEXT(drep_rf_) + theme_bw(base_size = 5) + 
                labs(x="MAGs",y="Different microbial species", linetype="") +
                guides(linetype = guide_legend(override.aes = list(size = 1))) +
                theme(legend.position = "none", axis.title.y = element_text(size = 7), axis.title.x = element_text(size = 7),
                      axis.text.y = element_text(size = 6), axis.text.x = element_text(size = 6),
                      panel.grid.minor = element_line(size = 0.15), panel.grid.major = element_line(size = 0.3),
                      plot.margin=grid::unit(c(1,5,1,5), "mm")) +
                scale_x_continuous(expand=c(0,0), limits=c(0,42500), breaks = c(0, 10000, 20000, 30000, 40000)) + 
                scale_y_continuous(expand=c(0,0), limits=c(0,42500), breaks = c(0, 10000, 20000, 30000, 40000)) +
                scale_fill_manual(values=c("#034e7b")) + scale_colour_manual(values=c("#034e7b"))

# Add customizations
drep_rf2 <- ggplot_build(drep_rf_plot)
drep_rf2$data[[1]]$size <- 3           # Point size
drep_rf2$data[[2]]$linewidth <- 0.8    # Line size
drep_rf2 <- ggplot_gtable(drep_rf2)
grid.draw(drep_rf2)
```


### Make Figure 2
```{r, include=TRUE}
core_gen <- readRDS("../data/MFD-LR/core_genera.rds") + labs(fill="MIMAG ranking:") + scale_x_discrete(guide = guide_axis(angle = 30)) + 
            theme(plot.margin = unit(c(2, 2, -3, 5), "mm"), legend.text=element_text(size=7), legend.title = element_text(size=7),
                  legend.box.spacing = unit(0, "mm"), legend.position = "bottom")

fig2_ab <- ggarrange(sankey, as.ggplot(drep_rf2), nrow=1, ncol=2, legend = "none", align = "h",
                     common.legend = FALSE, labels = c("a","b"), font.label = list(size = 9), widths=c(1,1))

fig2_cf <- ggarrange(plot_mag_size, plot_mag_cov, plot_mag_n50, plot_mag_cd, nrow=2, ncol=2, legend = "none", align = "hv",
                     common.legend = FALSE, labels = c("c","d","e","f"), font.label = list(size = 9)) 

fig2_cfg <- ggarrange(fig2_cf, core_gen, common.legend = FALSE, nrow=1, ncol=2, align = "v", labels = c(NA,"g"), font.label = list(size = 9), widths=c(1.15,1)) +
            theme(plot.margin = unit(c(0, 0, 5, -11), "mm"))

fig2 <- ggarrange(fig2_ab,fig2_cfg, nrow=2, ncol=1, legend = "none", align = "h", common.legend = FALSE, heights=c(1.25,1), widths=c(1,1))
fig2

ggsave(file="../analysis/figures/Fig2.pdf", height = 140, width = 180, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Plot species-level rarefaction curve, by habitat
```{r, include=TRUE}
drep_rf_hab <- aggregate(mags[mags$mfd_hab1_ != "Other",]$drep_cluster,
                         by=list(mags[mags$mfd_hab1_ != "Other",]$drep_cluster,
                                 mags[mags$mfd_hab1_ != "Other",]$mfd_hab1_), FUN=length)

drep_rf_hab_tmp <- split(drep_rf_hab$x, drep_rf_hab$Group.2)
drep_rf_hab_ <- iNEXT(drep_rf_hab_tmp, q=0, datatype="abundance")

drep_rf_hab_plot <- ggiNEXT(drep_rf_hab_, color.var="Assemblage") + theme_bw(base_size = 5) + 
                    labs(x="MAGs",y="Different microbial species", linetype="", fill="", color="") +
                    theme(legend.position = "bottom", strip.text = element_text(size=7), legend.spacing = unit(0.1, "mm"),
                          axis.title.y = element_text(size = 7), axis.title.x = element_text(size = 7),
                          axis.text.y = element_text(size = 6), axis.text.x = element_text(size = 6),
                          legend.text=element_text(size=7),legend.title = element_text(size=7),
                          legend.box.margin = margin(0, 0, 0, 0), legend.margin = margin(0, 0, 0, 0), 
                          panel.grid.minor = element_line(size = 0.15), panel.grid.major = element_line(size = 0.3)) + 
                    scale_color_manual(values=labels) + scale_fill_manual(values=labels) +
                    scale_shape_manual(values=c(19,19,19,19,19,19,19,19,19,19,19,19)) +
                    guides(color=guide_legend(override.aes = list(linewidth=NA, fill=NA, size=1.5)), shape="none",linetype = "none", fill = "none") +
                    theme(plot.margin = unit(c(2, 4, 0, 2), "mm"), legend.key.height=unit(0.4, 'cm'), legend.key.width=unit(0.25, 'cm'))

# Add customizations
drep_rf_hab_plot2 <- ggplot_build(drep_rf_hab_plot)
drep_rf_hab_plot2$data[[1]]$size <- 2           # Point size
drep_rf_hab_plot2$data[[2]]$linewidth <- 0.4    # Line size
drep_rf_hab_plot2 <- ggplot_gtable(drep_rf_hab_plot2)
grid.draw(drep_rf_hab_plot2)

drep_rf_hab_plot3 <- grid.arrange(drep_rf_hab_plot2)

ggsave(drep_rf_hab_plot3, file="../analysis/figures/FigS6.pdf", height = 120, width = 140, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Plot most abundant species clusters
```{r, include=TRUE}
drep_rf_n <- aggregate(mags$drep_cluster, by=list(mags$drep_cluster), FUN=length)
colnames(drep_rf_n) <- c("drep_cluster","drep_cluster_n")
mags_drep <- merge(mags[mags$rep_status == TRUE,], drep_rf_n, by="drep_cluster")

mags_drep_n <- mags_drep[, c("drep_cluster_n", "phylum", "family", "genus", "drep_cluster")]
mags_drep_n2 <- mags_drep_n[order(mags_drep_n$drep_cluster_n,decreasing=TRUE),]
mags_drep_n2 <- as.data.frame(head(mags_drep_n2,50))
mags_drep_n2$id <- paste(mags_drep_n2$genus,mags_drep_n2$drep_cluster,sep=" genus, cluster ")

plot_drep_n <- ggplot(data=mags_drep_n2, aes(x=factor(id,level=rev(id)), y=drep_cluster_n, color=phylum)) + 
               labs(title="", y="MAGs per cluster", x="", color="Phyla:") + 
               geom_segment(aes(xend=id), yend=0, size=0.25, color="gray70") + geom_point() + 
               scale_y_continuous(expand=c(0,0), limits=c(0,45), breaks = c(0, 10, 20, 30, 40)) +
               preset_gen2 + theme(legend.position = "right") + scale_color_brewer(palette = "Dark2") + coord_flip() 

plot_drep_n
ggsave(file="../analysis/figures/FigS7.pdf", height = 180, width = 160, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Plot species-level rarefaction curve, by most common phyla
```{r, include=TRUE}
mags_filt <- mags[mags$phylum == "Actinomycetota" | mags$phylum == "Pseudomonadota" | 
                  mags$phylum == "Acidobacteriota" | mags$phylum == "Chloroflexota",]

drep_rf_phyl <- aggregate(mags_filt$drep_cluster, by=list(mags_filt$drep_cluster,mags_filt$phylum), FUN=length)

drep_rf_phyl_tmp <- split(drep_rf_phyl$x, drep_rf_phyl$Group.2)
drep_rf_phyl_ <- iNEXT(drep_rf_phyl_tmp, q=0, datatype="abundance")

drep_rf_phyl_plot <- ggiNEXT(drep_rf_phyl_, color.var="Assemblage") + theme_bw(base_size = 5) + 
                     labs(x="MAGs",y="Different microbial species", linetype="") +
                     guides(linetype = guide_legend(override.aes = list(size = 1))) +
                     theme(legend.position = "none",strip.text = element_text(size=7),
                           axis.title.y = element_text(size = 7), axis.title.x = element_text(size = 7),
                           axis.text.y = element_text(size = 6), axis.text.x = element_text(size = 6),
                           panel.grid.minor = element_line(size = 0.15), panel.grid.major = element_line(size = 0.3))  +
                     scale_shape_manual(values=c(19,19,19,19)) + 
                     scale_color_manual(values=c("#7fc97f","#beaed4","#fdc086","#f0027f")) +
                     scale_fill_manual(values=c("#7fc97f","#beaed4","#fdc086","#f0027f")) + 
                     facet_rep_wrap(~Assemblage, ncol = 2, scales="free", repeat.tick.labels = TRUE) 

# Add customizations
drep_rf_phyl_plot2 <- ggplot_build(drep_rf_phyl_plot)
drep_rf_phyl_plot2$data[[1]]$size <- 3           # Point size
drep_rf_phyl_plot2$data[[2]]$linewidth <- 0.8    # Line size
drep_rf_phyl_plot2 <- ggplot_gtable(drep_rf_phyl_plot2)
grid.draw(drep_rf_phyl_plot2)

ggsave(drep_rf_phyl_plot2, file="../analysis/figures/FigS10.pdf", height = 145, width = 140, useDingbats=FALSE,  units ="mm", dpi = 300)
```


### Plot phylum-level (top 20) heatmap by habitat
```{r, include=TRUE}
# Count across all MAGs
sum_phylum <- as.data.frame(table(mags$phylum))
sum_phylum$Prop <- round(sum_phylum$Freq/sum(sum_phylum$Freq)*100,3)

phylum_ <- sum_phylum
phylum_ <- phylum_[order(phylum_$Freq,decreasing=TRUE),]
phylum_ <- as.data.frame(head(phylum_,20))

mags$phylum_ <- ifelse(mags$phylum %in% phylum_$Var1, mags$phylum, "Other phyla")

phylum_count_all <- aggregate(mags$phylum_, by=list(mags$phylum_,mags$mfd_hab1_), FUN=length, drop=FALSE)
phylum_count_all$x[is.na(phylum_count_all$x)] <- 0

phylum_count_hab <- aggregate(mags$mfd_hab1_, by=list(mags$mfd_hab1_), FUN=length)
colnames(phylum_count_hab) <- c("Group.2","mags_hab")

phylum_count_all <- merge(phylum_count_all,phylum_count_hab, by="Group.2")
phylum_count_all$frac <- round(phylum_count_all$x / phylum_count_all$mags_hab * 100,2)

# Count across dereplicated MAGs
level_phyla <- aggregate(mags[mags$rep_status == TRUE,]$phylum_, by=list(mags[mags$rep_status == TRUE,]$phylum_), FUN=length)
level_phyla <- level_phyla[order(level_phyla$x,decreasing=FALSE),]
level_phyla$mags_hab <- sum(level_phyla$x)
level_phyla$frac <- round(level_phyla$x / level_phyla$mags_hab * 100,2)
level_phyla$Group.2 <- "All habitats (dereplicated)"
level_phyla <- level_phyla %>% relocate(Group.2)
phylum_count_all2 <- rbind(phylum_count_all,level_phyla)

# Plot
phylum_count_all2$frac2 <- paste("(",phylum_count_all2$frac," %)", sep="")

plot_phylum <- ggplot(data = phylum_count_all2, mapping = aes(x = Group.2, y = factor(Group.1,level=level_phyla$Group.1), fill = frac)) +
               geom_tile(width=1) + scale_x_discrete(guide = guide_axis(angle = 45)) + labs(title="", y="", x="", fill="Fraction of MAGs (%):") +
               geom_text(aes(label = x), size=1.75, vjust=-0.7) + geom_text(aes(label = frac2), size=1.75, vjust=1) +
               scale_fill_gradient2(midpoint = 0.05, limits=c(0.01,50), breaks=c(0.01,0.1,1,10), low = "gray95", mid= "#fed976", high = "#800026", 
                                    trans="log10", labels=c("0.01","0.1","1","10"))  +
               guides(fill = guide_colorbar(title.position = "left", title.vjust = 0.75)) + preset_gen2

plot_phylum
ggsave(file="../analysis/figures/FigS11.pdf", height = 180, width = 140, useDingbats=FALSE,  units ="mm", dpi = 300)
```

